<!--#include virtual="/includes/starthtml.html" -->

<!-- HTML for http://www.broadinstitute.org/mpg/snpsea/ -->

<title>SNPsea</title>

<!--#include virtual="/includes/general.html" -->
<!--#include virtual="/includes/topbanner.html" -->
<!--#include virtual="/includes/nav_tools.html" -->

<link href='http://fonts.googleapis.com/css?family=Fira+Sans' rel='stylesheet' type='text/css'>

<style>
    blockquote {
        margin-left: 3em;
    }
    table { margin-left: 5em; }
    td { padding: 0.5em; }
    #content-inner p {
        padding: 0px 40px 0px 40px !important;
    }
    #content p, pre, ul, ol, li {
        font-family: 'Fira Sans', sans-serif;
        font-size: 12px !important;
    }
    #content h2 {
        font-size: 16px;
    }
</style>

<div id="breadcrumbs">
    <a href="/index.html">Home</a> &gt; <a href="/mpg">Medical
    &amp; Population Genetics</a> &gt; SNPsea
</div>

<div id="content">

    <h1 class="about">SNPsea: an algorithm to identify cell types and pathways
    affected by risk loci</h1>

    <p><script type="text/javascript">
<!--
h='&#98;&#114;&#x6f;&#x61;&#100;&#x69;&#110;&#x73;&#116;&#x69;&#116;&#x75;&#116;&#x65;&#46;&#x6f;&#114;&#x67;';a='&#64;';n='&#x73;&#108;&#x6f;&#x77;&#x69;&#x6b;&#x6f;&#x77;';e=n+a+h;
document.write('<a h'+'ref'+'="ma'+'ilto'+':'+e+'">'+e+'<\/'+'a'+'>');
// -->
    </script><noscript>slowikow at broadinstitute dot org</noscript></p>
  
    <h2>Overview</h2>

    <p>SNPsea is an algorithm to identify cell types and pathways likely to be
    affected by risk loci. It requires a list of SNP identifiers and a matrix
    of genes and conditions.</p>

    <p>Genome-wide association studies (GWAS) have discovered multiple genomic
    loci associated with risk for different types of disease. SNPsea provides
    a simple way to determine the types of cells influenced by genes in these
    risk loci.</p>
    
    <p>Suppose disease-associated alleles influence a small number of
    pathogenic cell types. We hypothesize that genes with critical functions
    in those cell types are likely to be within risk loci for that disease. We
    assume that a gene's specificity to a cell type is a reasonable indicator
    of its importance to the unique function of that cell type.</p>

    <p>First, we identify the genes in linkage disequilibrium (LD) with the
    given trait-associated SNPs and score the gene set for specificity to each
    cell type. Next, we define a null distribution of scores for each cell
    type by sampling random SNP sets matched on the number of linked genes.
    Finally, we evaluate the significance of the original gene set's
    specificity by comparison to the null distributions: we calculate an exact
    permutation p-value.</p>

    <p>SNPsea is a general algorithm. You may provide your own:</p>
    <ol style="list-style-type: decimal; margin-left: 5em;">
    <li>Continuous gene matrix with gene expression profiles (or other values).</li>
    <li>Binary gene annotation matrix with presence/absence 1/0 values.</li>
    </ol>

    <p>We provide three expression matrices and one annotation matrix.</p>

    <p>The columns of the matrix may be tissues, cell types, GO annotation
    codes, or other <em>conditions</em>. Continuous matrices <em>must</em> be
    normalized before running SNPsea: columns must be directly comparable to
    each other.</p>

    <h2>Citation</h2>

    <p>If you benefit from this method, please cite:</p>

    <blockquote>
        <p>Slowikowski, K et al. <strong>SNPsea: an algorithm to identify
        pathways, tissues, and conditions influenced by risk loci.</strong>
        Bioinformatics (2014). doi:<a href=
        "http://bioinformatics.oxfordjournals.org/content/early/2014/05/10/bioinformatics.btu326">10.1093/bioinformatics/btu326</a>
    </blockquote>

    <p>For additional information, please see:</p>

    <blockquote>
        <p>Hu, X et al. <strong>Integrating autoimmune risk loci
        with gene-expression data identifies specific pathogenic
        immune cell subsets.</strong> The American Journal of Human
        Genetics 89, 496-506 (2011). <a href=
        "http://www.ncbi.nlm.nih.gov/pubmed/21963258">PubMed</a></p>
    </blockquote>

<!--
    <h2>Run (Coming soon...)</h2>

    <p>You may submit a list of SNP identifiers or a file with SNP identifiers
    listed in a single column. Your job will run SNPsea on the Broad
    servers.</p>

    <form name="submitForm" action="snpsea.php" method="POST" enctype="multipart/form-data">
        <table style="padding-left:4em;">
            <tr>
                <td>SNP list or file:</td>
                <td><input type="file" name="snpListFile" size="32"><br>
<textarea name="snpList" rows="10" cols="32">rs123
rs42</textarea></td>
            </tr>
            <tr>
                <td>Email (required):</td>
                <td><input type="text" name="email" maxlength="100" size="32"></td>
            </tr>
            <tr><td>
                <input type="submit" name="submit" value="Submit" class="buttonSubmit">
            </td></tr>
        </table>
    </form>
-->

    <h2>Downloads</h2>

    <table>
        <tr>
            <td>Documentation:</td>
            <td>
            <a href= "http://snpsea.readthedocs.org/en/latest/">HTML</a>
            &nbsp;
            <a href="https://readthedocs.org/projects/snpsea/downloads/pdf/latest/">PDF</a>
            &nbsp;
            <a href="https://readthedocs.org/projects/snpsea/downloads/epub/latest/">Epub</a>
            </td>
        </tr>
        <tr>
            <td>Executable (Linux 64):</td>
            <td><a href= "https://github.com/slowkow/snpsea/releases">https://github.com/slowkow/snpsea/releases</a></td>
        </tr>
        <tr>
            <td>Data:</td>
            <td><a href="http://dx.doi.org/10.6084/m9.figshare.871430">http://dx.doi.org/10.6084/m9.figshare.871430</a></td>
        </tr>
        <tr>
            <td>Source code:</td>
            <td><a href= "https://github.com/slowkow/snpsea/">https://github.com/slowkow/snpsea/</a></td>
        </tr>
    </table>

    <p><a href= "http://www.broadinstitute.org/~slowikow">Kamil Slowikowski</a>
    wrote the code in <a href="/personal/soumya/">Soumya Raychaudhuri</a>'s
    <a href="http://immunogenomics.hms.harvard.edu/">lab</a> at the
    <a href=
    "http://www.brighamandwomens.org/research/depts/medicine/genetics/default.aspx">
    Division of Genetics, Brigham and Women's Hospital</a>,
    <a href="http://hms.%C3%82%C2%ADharvard.%C3%82%C2%ADedu">Harvard
    Medical School</a>, and the Broad Institute.</p>

    <table style="margin:0px auto !important; width:500px !important">
        <tr>
        <td><img src="images/hms.jpg" width="40" alt= "HMS Logo"></td>
        <td><img src="images/bwh.png" width="42" alt= "BWH Logo"></td>
        <td><img src="images/broad.jpg" width="47" alt= "Broad Logo"></td>
        </tr>
    </table>

</div>

<script>
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new
Date();a=s.createElement(o),
m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
})(window,document,'script','//www.google-analytics.com/analytics.js','ga');
ga('create', 'UA-15531613-13', 'auto');
ga('send', 'pageview');
</script>

<!--#include virtual="/includes/quicklinks.html" -->
<!--#include virtual="/includes/footer.html" -->
<!--#include virtual="/includes/endhtml.html" -->
