<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta name="generator" content="pandoc">
            <title></title>

	<!-- Google WebFonts
  ================================================== -->
  <link href='http://fonts.googleapis.com/css?family=Droid+Sans' rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=Ubuntu+Mono' rel='stylesheet' type='text/css'>

	<!-- Mobile Specific Metas
  ================================================== -->
	<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">

    <!--[if lt IE 9]>
        <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

<style type="text/css">
/*
* Skeleton V1.2
* Copyright 2011, Dave Gamache
* www.getskeleton.com
* Free to use under the MIT license.
* http://www.opensource.org/licenses/mit-license.php
* 6/20/2012
*/


/** base.css **/

/* Table of Content
==================================================
	#Reset & Basics
	#Basic Styles
	#Site Styles
	#Typography
	#Links
	#Lists
	#Images
	#Buttons
	#Forms
	#Misc */

/* #Reset & Basics (Inspired by E. Meyers)
================================================== */
	html, body, div, span, applet, object, iframe, h1, h2, h3, h4, h5, h6, p, blockquote, pre, a, abbr, acronym, address, big, cite, code, del, dfn, em, img, ins, kbd, q, s, samp, small, strike, strong, sub, sup, tt, var, b, u, i, center, dl, dt, dd, ol, ul, li, fieldset, form, label, legend, table, caption, tbody, tfoot, thead, tr, th, td, article, aside, canvas, details, embed, figure, figcaption, footer, header, hgroup, menu, nav, output, ruby, section, summary, time, mark, audio, video {
		margin: 0;
		padding: 0;
		border: 0;
		font-size: 100%;
		font: inherit;
		vertical-align: baseline; }
	article, aside, details, figcaption, figure, footer, header, hgroup, menu, nav, section {
		display: block; }
    nav { 
        /*position: fixed;*/
        padding-top: 1em; }
	body {
		line-height: 1; }
	ol, ul {
		list-style: none; }
	blockquote, q {
		quotes: none; }
	blockquote:before, blockquote:after,
	q:before, q:after {
		content: '';
		content: none; }
	table {
		border-collapse: collapse;
		border-spacing: 0; }


/* #Basic Styles
================================================== */
	body {
		/*background: #fff;*/
		background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAC4AAAAXCAMAAABDEo1gAAAALVBMVEXu7u7z8/Pt7e3r6+vv7+/x8fHs7Ozw8PDo6Ojq6urp6eny8vL09PT19fX29vbJGhcbAAAB5klEQVR42k2TW4IkIQgEUUQLofb+x91At2b2x1eFaZJ0i7TWdTQTnastfaa1ob4lQraPlmtvy5z7nGvR7h5vyjMDfjOKF+2lYk1UlnjRiAh0X12SeTR4Lx2D1s48Z9Rjl0ZEZGvL0HloWZyin4mTzOz76Y31R6dwN7LJ3tCd8/LTGV0zl47pnScv3eBiYIa3cdt1+uWn1TGu6wH4S8cIobBaYrtT97r8o+htaj843KFld+EuG0ZE4+WrkufNba6Tr/3QruKH58VAxzImfHIX6qsD+gQrNmRCVlbF06kw+JapeyT0l0m1i+aEmBZvgoB115PPCPO5PnpDR4vpqyXVoZxv872K1Bq1/5eQznKyiu/GDE+SBR5+ROj+6BxYq0b5oKxnivIBAYdekHL8yEc3eXh6eVeOTR9BNdCHZpTMUB/RLo1lmy4LO2UdXvQE3W/Yau/bH/noMZHsvjUqqt7SxHBkR06XyVzMv5ncpPRkiusgGSt9FntmmjrcR1PShDdc0BLD2BLWuBhR1zPHvyKKdjKx0+PjIA1XgrJd/1OF/taCYqHJ7g3Xo6fQqh3cWbHHwukv9LPaCy2sjxnUjjaD+DPy8OvyydcK4v4Gi65uNuiZf5rgINhdfb9/DLF+aS/6sXjtRiB/AetLF9OdGDz6AAAAAElFTkSuQmCC);
		/*font: 18px/27px "Droid Sans", sans-serif; /*"HelveticaNeue", "Helvetica Neue", Helvetica, Arial, sans-serif;*/
		font: 16px/24px "Droid Sans", sans-serif; /*"HelveticaNeue", "Helvetica Neue", Helvetica, Arial, sans-serif;*/
		color: #222;
		-webkit-font-smoothing: antialiased; /* Fix for webkit rendering */
		-webkit-text-size-adjust: 100%;
 }


/* #Typography
================================================== */
	h1, h2, h3, h4, h5, h6 {
		color: #333;
		/*font-family: "Georgia", "Times New Roman", serif;*/
		font-family: 'Droid Sans', sans-serif;
		text-decoration: none;
		/*font-family: "HelveticaNeue", "Helvetica Neue", Helvetica, Arial, sans-serif;*/
		font-weight: bold; }
	h1 a, h2 a, h3 a, h4 a, h5 a, h6 a { font-weight: inherit;
	    text-decoration: none; }
    h1 { font-size: 36px; line-height: 52px; margin-bottom: 24px;
        margin-left: -0.5em; display: block; border-bottom: 1px solid #bbb; }
    h1:nth-child(n+3) { margin-top: 2em; }
	h2 { font-size: 32px; line-height: 42px; margin-bottom: 20px;
        margin-top: 2em; margin-left: -0.3em }
	h3 { font-size: 23px; line-height: 36px; margin-bottom: 18px;
        margin-top: 20px; }
	h4 { font-size: 21px; line-height: 32px; margin-bottom: 16px; }
	h5 { font-size: 17px; line-height: 26px; margin-bottom: 12px; }
	h6 { font-size: 14px; line-height: 23px; margin-bottom: 8px; }
	.subheader { color: #333; }
    .title { font-size: 46px; line-height: 52px; margin-bottom: 24px;
        margin-top: 1em; display: block; border-bottom: 1px solid #bbb; }

    header table { width: 100%; }
    header table td { text-align: center; }

	p { margin: 0 0 1em 0; }
	p img { margin: 0; }
	p.lead { font-size: 21px; line-height: 27px; color: #333;  }

	em { font-style: italic; }
	strong { font-weight: bold; color: #333; }
	small { font-size: 80%; }

/*	Blockquotes  */
    blockquote, blockquote p { font-size: 16px; line-height: 24px; color: #333; }
	blockquote { margin: 0 0 20px; padding: 9px 20px 0 19px; border-left: 10px solid #ddd; }
	blockquote cite { display: block; font-size: 12px; color: #555; }
	blockquote cite:before { content: "\2014 \0020"; }
	blockquote cite a, blockquote cite a:visited, blockquote cite a:visited { color: #555; }

	hr { border: solid #ddd; border-width: 1px 0 0; clear: both; margin: 10px 0 30px; height: 0; }
    pre {
        padding: 0.5em 1em 0.5em 1em;
		font: 16px/20px "Ubuntu Mono", monospace; /*"HelveticaNeue", "Helvetica Neue", Helvetica, Arial, sans-serif;*/
        margin-bottom: 1em;
        display: block;
        overflow: auto;
        box-shadow: 0 0 10px rgba(0,0,0,.1);
        background-color: #fafafa; }


/* #Links
================================================== */
    a[href ^= "http"]:after { content: "⇗"; }
	a, a:visited { color: #333; text-decoration: underline; outline: 0; }
	a:hover, a:focus { color: #A60000; }
	p a, p a:visited { line-height: inherit; }


/* #Lists
================================================== */
	ul, ol { margin-bottom: 20px; }
	ul { list-style: none outside; }
	ol { list-style: decimal; }
	ol, ul.square, ul.circle, ul.disc { margin-left: 30px; }
	ul.square { list-style: square outside; }
	ul.circle { list-style: circle outside; }
	ul.disc { list-style: disc outside; }
	ul ul, ul ol,
	ol ol, ol ul { margin: 4px 0 5px 30px; font-size: 90%;  }
	ul ul li, ul ol li,
	ol ol li, ol ul li { margin-bottom: 6px; }
	li { line-height: 18px; margin-bottom: 12px; }
	ul.large li { line-height: 21px; }
	li p { line-height: 21px; }

/* #Images
================================================== */

	img {
        box-shadow: 0 0 10px rgba(0,0,0,.1);
		max-width: 100%;
		height: auto; }

/* #Buttons
================================================== */

	.button,
	button,
	input[type="submit"],
	input[type="reset"],
	input[type="button"] {
		background: #eee; /* Old browsers */
		background: #eee -moz-linear-gradient(top, rgba(255,255,255,.2) 0%, rgba(0,0,0,.2) 100%); /* FF3.6+ */
		background: #eee -webkit-gradient(linear, left top, left bottom, color-stop(0%,rgba(255,255,255,.2)), color-stop(100%,rgba(0,0,0,.2))); /* Chrome,Safari4+ */
		background: #eee -webkit-linear-gradient(top, rgba(255,255,255,.2) 0%,rgba(0,0,0,.2) 100%); /* Chrome10+,Safari5.1+ */
		background: #eee -o-linear-gradient(top, rgba(255,255,255,.2) 0%,rgba(0,0,0,.2) 100%); /* Opera11.10+ */
		background: #eee -ms-linear-gradient(top, rgba(255,255,255,.2) 0%,rgba(0,0,0,.2) 100%); /* IE10+ */
		background: #eee linear-gradient(top, rgba(255,255,255,.2) 0%,rgba(0,0,0,.2) 100%); /* W3C */
	  border: 1px solid #aaa;
	  border-top: 1px solid #ccc;
	  border-left: 1px solid #ccc;
	  -moz-border-radius: 3px;
	  -webkit-border-radius: 3px;
	  border-radius: 3px;
	  color: #444;
	  display: inline-block;
	  font-size: 11px;
	  font-weight: bold;
	  text-decoration: none;
	  text-shadow: 0 1px rgba(255, 255, 255, .75);
	  cursor: pointer;
	  margin-bottom: 20px;
	  line-height: normal;
	  padding: 8px 10px;
      font-family: "Droid Sans", sans-serif; /*"HelveticaNeue", "Helvetica Neue", Helvetica, Arial, sans-serif;*/ }

	.button:hover,
	button:hover,
	input[type="submit"]:hover,
	input[type="reset"]:hover,
	input[type="button"]:hover {
		color: #222;
		background: #ddd; /* Old browsers */
		background: #ddd -moz-linear-gradient(top, rgba(255,255,255,.3) 0%, rgba(0,0,0,.3) 100%); /* FF3.6+ */
		background: #ddd -webkit-gradient(linear, left top, left bottom, color-stop(0%,rgba(255,255,255,.3)), color-stop(100%,rgba(0,0,0,.3))); /* Chrome,Safari4+ */
		background: #ddd -webkit-linear-gradient(top, rgba(255,255,255,.3) 0%,rgba(0,0,0,.3) 100%); /* Chrome10+,Safari5.1+ */
		background: #ddd -o-linear-gradient(top, rgba(255,255,255,.3) 0%,rgba(0,0,0,.3) 100%); /* Opera11.10+ */
		background: #ddd -ms-linear-gradient(top, rgba(255,255,255,.3) 0%,rgba(0,0,0,.3) 100%); /* IE10+ */
		background: #ddd linear-gradient(top, rgba(255,255,255,.3) 0%,rgba(0,0,0,.3) 100%); /* W3C */
	  border: 1px solid #888;
	  border-top: 1px solid #aaa;
	  border-left: 1px solid #aaa; }

	.button:active,
	button:active,
	input[type="submit"]:active,
	input[type="reset"]:active,
	input[type="button"]:active {
		border: 1px solid #666;
		background: #ccc; /* Old browsers */
		background: #ccc -moz-linear-gradient(top, rgba(255,255,255,.35) 0%, rgba(10,10,10,.4) 100%); /* FF3.6+ */
		background: #ccc -webkit-gradient(linear, left top, left bottom, color-stop(0%,rgba(255,255,255,.35)), color-stop(100%,rgba(10,10,10,.4))); /* Chrome,Safari4+ */
		background: #ccc -webkit-linear-gradient(top, rgba(255,255,255,.35) 0%,rgba(10,10,10,.4) 100%); /* Chrome10+,Safari5.1+ */
		background: #ccc -o-linear-gradient(top, rgba(255,255,255,.35) 0%,rgba(10,10,10,.4) 100%); /* Opera11.10+ */
		background: #ccc -ms-linear-gradient(top, rgba(255,255,255,.35) 0%,rgba(10,10,10,.4) 100%); /* IE10+ */
		background: #ccc linear-gradient(top, rgba(255,255,255,.35) 0%,rgba(10,10,10,.4) 100%); /* W3C */ }

	.button.full-width,
	button.full-width,
	input[type="submit"].full-width,
	input[type="reset"].full-width,
	input[type="button"].full-width {
		width: 100%;
		padding-left: 0 !important;
		padding-right: 0 !important;
		text-align: center; }

	/* Fix for odd Mozilla border & padding issues */
	button::-moz-focus-inner,
	input::-moz-focus-inner {
    border: 0;
    padding: 0;
	}

/* #Forms
================================================== */

	form {
		margin-bottom: 20px; }
	fieldset {
		margin-bottom: 20px; }
	input[type="text"],
	input[type="password"],
	input[type="email"],
	textarea,
	select {
		border: 1px solid #ccc;
		padding: 6px 4px;
		outline: none;
		-moz-border-radius: 2px;
		-webkit-border-radius: 2px;
		border-radius: 2px;
      font: 13px "Droid Sans", sans-serif; /*"HelveticaNeue", "Helvetica Neue", Helvetica, Arial, sans-serif;*/ }
		color: #333;
		margin: 0;
		width: 210px;
		max-width: 100%;
		display: block;
		margin-bottom: 20px;
		background: #fff; }
	select {
		padding: 0; }
	input[type="text"]:focus,
	input[type="password"]:focus,
	input[type="email"]:focus,
	textarea:focus {
		border: 1px solid #aaa;
 		color: #444;
 		-moz-box-shadow: 0 0 3px rgba(0,0,0,.2);
		-webkit-box-shadow: 0 0 3px rgba(0,0,0,.2);
		box-shadow:  0 0 3px rgba(0,0,0,.2); }
	textarea {
		min-height: 60px; }
	label,
	legend {
		display: block;
		font-weight: bold;
		font-size: 13px;  }
	select {
		width: 220px; }
	input[type="checkbox"] {
		display: inline; }
	label span,
	legend span {
		font-weight: normal;
		font-size: 13px;
		color: #444; }

/* #Misc
================================================== */
	.remove-bottom { margin-bottom: 0 !important; }
	.half-bottom { margin-bottom: 10px !important; }
	.add-bottom { margin-bottom: 20px !important; }


/** skeleton.css **/


/* Table of Contents
==================================================
    #Base 960 Grid
    #Tablet (Portrait)
    #Mobile (Portrait)
    #Mobile (Landscape)
    #Clearing */

/* #Base 960 Grid
================================================== */

    .container                                  { position: relative; width: 960px; margin: 0 auto; padding: 0; }
    .container .column,
    .container .columns                         { float: left; display: inline; margin-left: 10px; margin-right: 10px; }
    .row                                        { margin-bottom: 20px; }

    /* Nested Column Classes */
    .column.alpha, .columns.alpha               { margin-left: 0; }
    .column.omega, .columns.omega               { margin-right: 0; }

    /* Base Grid */
    .container .one.column,
    .container .one.columns                     { width: 40px;  }
    .container .two.columns                     { width: 100px; }
    .container .three.columns                   { width: 160px; }
    .container .four.columns                    { width: 220px; }
    .container .five.columns                    { width: 280px; }
    .container .six.columns                     { width: 340px; }
    .container .seven.columns                   { width: 400px; }
    .container .eight.columns                   { width: 460px; }
    .container .nine.columns                    { width: 520px; }
    .container .ten.columns                     { width: 580px; }
    .container .eleven.columns                  { width: 640px; }
    .container .twelve.columns                  { width: 700px; }
    .container .thirteen.columns                { width: 760px; }
    .container .fourteen.columns                { width: 820px; }
    .container .fifteen.columns                 { width: 880px; }
    .container .sixteen.columns                 { width: 940px; }

    .container .one-third.column                { width: 300px; }
    .container .two-thirds.column               { width: 620px; }

    /* Offsets */
    .container .offset-by-one                   { padding-left: 60px;  }
    .container .offset-by-two                   { padding-left: 120px; }
    .container .offset-by-three                 { padding-left: 180px; }
    .container .offset-by-four                  { padding-left: 240px; }
    .container .offset-by-five                  { padding-left: 300px; }
    .container .offset-by-six                   { padding-left: 360px; }
    .container .offset-by-seven                 { padding-left: 420px; }
    .container .offset-by-eight                 { padding-left: 480px; }
    .container .offset-by-nine                  { padding-left: 540px; }
    .container .offset-by-ten                   { padding-left: 600px; }
    .container .offset-by-eleven                { padding-left: 660px; }
    .container .offset-by-twelve                { padding-left: 720px; }
    .container .offset-by-thirteen              { padding-left: 780px; }
    .container .offset-by-fourteen              { padding-left: 840px; }
    .container .offset-by-fifteen               { padding-left: 900px; }



/* #Tablet (Portrait)
================================================== */

    /* Note: Design for a width of 768px */

    @media only screen and (min-width: 768px) and (max-width: 959px) {
        .container                                  { width: 768px; }
        .container .column,
        .container .columns                         { margin-left: 10px; margin-right: 10px;  }
        .column.alpha, .columns.alpha               { margin-left: 0; margin-right: 10px; }
        .column.omega, .columns.omega               { margin-right: 0; margin-left: 10px; }
        .alpha.omega                                { margin-left: 0; margin-right: 0; }

        .container .one.column,
        .container .one.columns                     { width: 28px; }
        .container .two.columns                     { width: 76px; }
        .container .three.columns                   { width: 124px; }
        .container .four.columns                    { width: 172px; }
        .container .five.columns                    { width: 220px; }
        .container .six.columns                     { width: 268px; }
        .container .seven.columns                   { width: 316px; }
        .container .eight.columns                   { width: 364px; }
        .container .nine.columns                    { width: 412px; }
        .container .ten.columns                     { width: 460px; }
        .container .eleven.columns                  { width: 508px; }
        .container .twelve.columns                  { width: 556px; }
        .container .thirteen.columns                { width: 604px; }
        .container .fourteen.columns                { width: 652px; }
        .container .fifteen.columns                 { width: 700px; }
        .container .sixteen.columns                 { width: 748px; }

        .container .one-third.column                { width: 236px; }
        .container .two-thirds.column               { width: 492px; }

        /* Offsets */
        .container .offset-by-one                   { padding-left: 48px; }
        .container .offset-by-two                   { padding-left: 96px; }
        .container .offset-by-three                 { padding-left: 144px; }
        .container .offset-by-four                  { padding-left: 192px; }
        .container .offset-by-five                  { padding-left: 240px; }
        .container .offset-by-six                   { padding-left: 288px; }
        .container .offset-by-seven                 { padding-left: 336px; }
        .container .offset-by-eight                 { padding-left: 384px; }
        .container .offset-by-nine                  { padding-left: 432px; }
        .container .offset-by-ten                   { padding-left: 480px; }
        .container .offset-by-eleven                { padding-left: 528px; }
        .container .offset-by-twelve                { padding-left: 576px; }
        .container .offset-by-thirteen              { padding-left: 624px; }
        .container .offset-by-fourteen              { padding-left: 672px; }
        .container .offset-by-fifteen               { padding-left: 720px; }
    }


/*  #Mobile (Portrait)
================================================== */

    /* Note: Design for a width of 320px */

    @media only screen and (max-width: 767px) {
        .container { width: 300px; }
        .container .columns,
        .container .column { margin: 0; }

        .container .one.column,
        .container .one.columns,
        .container .two.columns,
        .container .three.columns,
        .container .four.columns,
        .container .five.columns,
        .container .six.columns,
        .container .seven.columns,
        .container .eight.columns,
        .container .nine.columns,
        .container .ten.columns,
        .container .eleven.columns,
        .container .twelve.columns,
        .container .thirteen.columns,
        .container .fourteen.columns,
        .container .fifteen.columns,
        .container .sixteen.columns,
        .container .one-third.column,
        .container .two-thirds.column  { width: 300px; }

        /* Offsets */
        .container .offset-by-one,
        .container .offset-by-two,
        .container .offset-by-three,
        .container .offset-by-four,
        .container .offset-by-five,
        .container .offset-by-six,
        .container .offset-by-seven,
        .container .offset-by-eight,
        .container .offset-by-nine,
        .container .offset-by-ten,
        .container .offset-by-eleven,
        .container .offset-by-twelve,
        .container .offset-by-thirteen,
        .container .offset-by-fourteen,
        .container .offset-by-fifteen { padding-left: 0; }

    }


/* #Mobile (Landscape)
================================================== */

    /* Note: Design for a width of 480px */

    @media only screen and (min-width: 480px) and (max-width: 767px) {
        .container { width: 420px; }
        .container .columns,
        .container .column { margin: 0; }

        .container .one.column,
        .container .one.columns,
        .container .two.columns,
        .container .three.columns,
        .container .four.columns,
        .container .five.columns,
        .container .six.columns,
        .container .seven.columns,
        .container .eight.columns,
        .container .nine.columns,
        .container .ten.columns,
        .container .eleven.columns,
        .container .twelve.columns,
        .container .thirteen.columns,
        .container .fourteen.columns,
        .container .fifteen.columns,
        .container .sixteen.columns,
        .container .one-third.column,
        .container .two-thirds.column { width: 420px; }
    }

    @media only screen and (min-width: 768px) {
        nav { position: fixed; }
    }

/* #Clearing
================================================== */

    /* Self Clearing Goodness */
    .container:after { content: "\0020"; display: block; height: 0; clear: both; visibility: hidden; }

    /* Use clearfix class on parent to clear nested columns,
    or wrap each row of columns in a <div class="row"> */
    .clearfix:before,
    .clearfix:after,
    .row:before,
    .row:after {
      content: '\0020';
      display: block;
      overflow: hidden;
      visibility: hidden;
      width: 0;
      height: 0; }
    .row:after,
    .clearfix:after {
      clear: both; }
    .row,
    .clearfix {
      zoom: 1; }

    /* You can also use a <br class="clear" /> to clear columns */
    .clear {
      clear: both;
      display: block;
      overflow: hidden;
      visibility: hidden;
      width: 0;
      height: 0;
    }

</style>
    <style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
    </style>
<script src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" type="text/javascript"></script>
</head>
<body>
<div class="container">
    <div class="twelve columns offset-by-one content">
<h1 id="supplementary-notes">Supplementary Notes</h1>
<h2 id="algorithm-details">Algorithm Details</h2>
<p>For a given set of SNPs associated to a single trait, SNPsea tests whether implicated genes, in aggregate, are enriched for specificity to an annotation or condition in a user-provided matrix of genes and annotations/conditions. The algorithm consists of three steps:</p>
<ul class="incremental">
<li><p><strong>Step 1: Assigning genes to each SNP</strong></p>
<ul class="incremental">
<li>We use linkage disequilibrium (LD) to identify the genes implicated by each SNP.</li>
</ul></li>
<li><p><strong>Step 2: Calculating specificity scores</strong></p>
<ul class="incremental">
<li>We look up implicated genes in a user-provided matrix and calculate a specificity score for each annotation/condition based on the values of these genes.</li>
</ul></li>
<li><p><strong>Step 3: Testing significance</strong></p>
<ul class="incremental">
<li>We compare the specificity scores to scores obtained with random sets of matched SNP sets and use those scores to calculate an empirical p-value.</li>
</ul></li>
</ul>
<h3 id="step-1-assigning-genes-to-each-snp" class="unnumbered">Step 1: Assigning genes to each SNP</h3>
<p>Accurate analyses must address the critical issue that SNPs infrequently implicate a locus with a single gene. Instead, SNPs implicate loci in LD with 0 to 20 genes that may plausibly influence the associated trait (<strong>Supplementary Figure 2</strong>). We determine the genes implicated by each SNP using a previously described strategy (<strong>Supplementary Figure 1</strong> and <span class="citation">@rossin_proteins_2011</span>). First, we define a linkage interval that spans between the furthest neighboring SNPs within a 1 Mb window with <span class="math">\(r^{2}&gt;0.5\)</span> using European reference genomes <span class="citation">@consortium_integrated_2012</span>. Next, we extend each interval to the nearest recombination hotspots with recombination rate &gt;3 cM/Mb <span class="citation">@myers_fine-scale_2005</span>. To address the case when no genes overlap an interval, we provide an option to extend the interval up- and downstream (by default 10 Kb). After each SNP has been assigned an interval and a set of overlapping genes, we merge SNPs with shared genes into a single locus to avoid multiple-counting of genes.</p>
<p>By default, we assume that one gene in each associated locus is associated with the given trait. However, we expect some loci with multiple genes due to poorly mapped regions in the genome, or due to regions with high LD that span long stretches across the chromosome. We also allow investigators the alternative to assume that all genes within a locus are associated, and we compare results of the two alternatives with four phenotypes (<strong>Supplementary Figure 4</strong>). [par:single<sub>t</sub>otal]</p>
<ol class="incremental" style="list-style-type: decimal">
<li><p>The <code>’--score single’</code> method (recommended) assumes that a single gene in each locus is associated with the given phenotype. For each condition, we choose the gene in each locus with the greatest specificity to that condition. We assume specificity is a good way to represent a gene’s importance to the function of a condition<em> </em>(cell type or tissue).</p></li>
<li><p>The <code>’--score total’</code> method assumes that all genes in a SNP’s linkage interval are associated. We account for all linked genes when calculating scores.</p></li>
</ol>
<h3 id="step-2-calculating-specificity-scores" class="unnumbered">Step 2: Calculating specificity scores</h3>
<p>SNPsea accepts matrices with continuous or binary values in each cell, but the algorithm for a continuous matrix differs from that for a binary matrix. Before running SNPsea, a matrix with continuous values must be normalized so that columns are comparable. <em>It is not appropriate to use this method on a “raw” matrix of expression values</em>, because the values across conditions (cell types, tissues, experiments) are not directly comparable.</p>
<h4 id="calculating-specificity-for-a-gene-expression-matrix">Calculating specificity for a gene expression matrix</h4>
<p><span class="math">\[A=\text{continuous gene expression matrix}\]</span></p>
<p>We row-normalize the matrix <span class="math">\(A\)</span> with <span class="math">\(m\)</span> genes and <span class="math">\(n\)</span> conditions by the L2 norm. This results in a new matrix <span class="math">\(A&#39;\)</span> where each value is between 0 and 1 and represents specificity of gene <span class="math">\(i\)</span> in condition <span class="math">\(j\)</span>. A value of 1 indicates complete specificity to condition <span class="math">\(j\)</span> and 0 indicates complete absence in condition <span class="math">\(j\)</span> as described previously <span class="citation">@hu_integrating_2011</span>.</p>
<p><span class="math">\[A&#39;_{i,j}=\dfrac{A_{i,j}}{\sqrt{A_{i,1}^{2}+A_{i,2}^{2}+\cdots+A_{i,n}^{2}}}\]</span></p>
<p>Next, we rank each column by ascending percentile (lower rank indicates greater specificity) to produce a new matrix <span class="math">\(A&#39;&#39;\)</span>. Each value is between 0 and 1 and represents a nonparametric rank of specificity to condition <span class="math">\(j\)</span> for gene <span class="math">\(i\)</span>:</p>
<p><span class="math">\[A&#39;&#39;_{i,j}=\dfrac{\text{Rank}(A&#39;_{i,j})}{m}\]</span></p>
<p>In each locus-condition pair, we choose the single gene with the highest specificity to that condition. To correct for testing multiple genes <span class="math">\(g_{k}\)</span> in each SNP locus’ set of genes <span class="math">\(k\)</span>, we create a new matrix <span class="math">\(P\)</span> of probabilities. It consists of one row for each SNP locus <span class="math">\(k\)</span> and one column for each condition <span class="math">\(j\)</span>. We correct for testing multiple genes in each SNP locus, because we choose the single gene with the greatest specificity among <span class="math">\(g_{k}\)</span> genes in the locus. We also provide the option to use all genes in each SNP locus <span class="math">\(i\in k\)</span>.</p>
<table style="margin:auto;text-align:center" cellpadding="20">
<tr>
<td>
<code>’--score single’</code>
</td><td>
<code>’--score total’</code>
</td>
</tr>
<tr>
<td>
<span class="math">\(P_{k,j}=1-(1-\text{Min}(A&#39;&#39;_{k,j}))^{g_{k}}\)</span>
</td>
<td>
<span class="math">\(P_{k,j}=-\sum_{i\in k}\text{log}A&#39;&#39;_{i,j}\)</span>
</td>
</tr>
</table>

<h4 id="calculating-specificity-for-a-binary-presenceabsence-matrix">Calculating specificity for a binary presence/absence matrix</h4>
<p><span class="math">\[B=\text{binary (1/0) presence/absence matrix}\]</span></p>
<p>Matrix <span class="math">\(B\)</span> with <span class="math">\(m\)</span> genes and <span class="math">\(n\)</span> conditions has binary values, so we do not transform it into <span class="math">\(B&#39;\)</span> or <span class="math">\(B&#39;&#39;\)</span>. Instead, we create a matrix <span class="math">\(P\)</span> of hypergeometric probabilities. Below, <span class="math">\(m_{j}\)</span> genes are present in condition <span class="math">\(j\)</span>, <span class="math">\(g_{k}\)</span> genes are in locus <span class="math">\(k\)</span> and <span class="math">\(g_{k,j}\)</span> of them are present in condition <span class="math">\(j\)</span>. By default, we account for the presence or absence of any of the locus’ genes in condition <span class="math">\(j\)</span>. We also provide the option to account for the number of genes in each locus.</p>
<table style="margin:auto;text-align:center" cellpadding="20">
<tr><td></td><td>
<code>’--score single’</code>
</td> <td>
<code>’--score total’</code>
</td></tr>
<tr>
<td>
<span class="math">\(p(x)=\dfrac{{m_{j} \choose x}{m-m_{j} \choose g_{k}-x}}{{m \choose g_{k}}}\)</span>
</td>
<td>
<span class="math">\(P_{k,j}=\begin{cases} 1-p(0) &amp; g_{k,j}&gt;0\\ 1 &amp; g_{k,j}=0 \end{cases}\)</span>
</td><td>
<span class="math">\(P_{k,j}=\begin{cases} 1-\sum_{x=0}^{g_{k,j}-1}p(x) &amp; g_{k,j}&gt;0\\ 1 &amp; g_{k,j}=0 \end{cases}\)</span>
</td>
</tr>
</table>

<p>Finally, we define a specificity score <span class="math">\(S_{j}\)</span> for each condition <span class="math">\(j\)</span> as the aggregate of all loci:</p>
<p><span class="math">\[S_{j}=-\sum_{k}\text{log}P_{k,j}\]</span></p>
<h3 id="step-3-testing-significance" class="unnumbered">Step 3: Testing significance</h3>
<h4 id="analytical-p-values">Analytical p-values</h4>
<p>We previously found that the use of analytical distributions such as the Poisson distribution can result in inaccurate p-values <span class="citation">@hu_integrating_2011</span>. SNP linkage intervals, gene densities, gene sizes and gene functions are correlated across the genome and are challenging to model analytically. This motivated us to use a sampling strategy instead.</p>
<h4 id="permutation-p-values">Permutation p-values</h4>
<p>For each condition, we use a sampling approach to calculate an empirical p-value: the tail probability of observing a condition-specificity score <span class="math">\(S_{j}\)</span> in a distribution of scores obtained with null SNP sets. We compute specificity scores <span class="math">\(S\)</span> for random SNP sets where each SNP is matched to a SNP in the user’s set on the number of linked genes. To adequately cover genes across the genome, we sample SNP sets from a list of LD-pruned SNPs that is a subset of the SNPs from the 1000 Genomes Project <span class="citation">@lango_allen_hundreds_2010</span>. For each condition <span class="math">\(j\)</span>, we calculate an exact permutation p-value <span class="citation">@phipson_permutation_2010</span> where <span class="math">\(a_{j}\)</span> is the number of sampled SNP sets and <span class="math">\(b_{j}\)</span> of them have a specificity score greater than or equal to the user’s score <span class="math">\(S_{j}\)</span>:</p>
<p><span class="math">\[p_{j}=\dfrac{b_{j}+1}{a_{j}+1}\]</span></p>
<p>We implemented adaptive sampling to calculate p-values more efficiently. As each condition is tested, we perform more iterations to increase resolution of significant p-values and fewer iterations for less significant p-values to save computation time. Two options allow the user to control the adaptive sampling:</p>
<ol class="incremental" style="list-style-type: decimal">
<li><p><code>’--max-iterations N’</code> The maximum number of iterations for each condition. We stop testing a condition after sampling <span class="math">\(N\)</span> SNP sets.</p></li>
<li><p><code>’--min-observations N’</code> The minimum number of observed higher scores <span class="math">\(S_{j}\)</span> required to stop sampling SNP sets for a condition <span class="math">\(j\)</span>.</p></li>
</ol>
<h2 id="data">Data</h2>
<p>Please find the data required to reproduce this analysis here: <a href="http://dx.doi.org/10.6084/m9.figshare.871430">http://dx.doi.org/10.6084/m9.figshare.871430</a></p>
<h3 id="gene-atlas-gene-expression-matrix" class="unnumbered">Gene Atlas gene expression matrix</h3>
<p>We downloaded the data from BioGPS: <a href="http://plugins.biogps.org/download/gnf1h-gcrma.zip">http://plugins.biogps.org/download/gnf1h-gcrma.zip</a></p>
<p>We averaged the expression values for tissue replicates. For each gene, we selected the single probe with the largest minimum value. Finally, we converted the file to <a href="http://www.broadinstitute.org/cancer/software/genepattern/gp_guides/file-formats/sections/gct">GCT format</a>.</p>
<h3 id="gene-ontology-binary-presenceabsence-matrix" class="unnumbered">Gene Ontology binary presence/absence matrix</h3>
<p>We downloaded the OBO file from Gene Ontology (data-version: 2013-06-29, CVS revision: 9700):</p>
<p><a href="http://www.geneontology.org">http://www.geneontology.org</a></p>
<p>For each gene, we climbed the hierarchy of ontology terms and applied parental terms. If a gene is annotated with some term <span class="math">\(T\)</span>, we also add all of the terms that are parents of <span class="math">\(T\)</span>. We copy terms between homologous genes using Homologene data (<a href="http://www.ncbi.nlm.nih.gov/homologene">http://www.ncbi.nlm.nih.gov/homologene</a>). If a mouse gene is annotated with some term and the human homolog is not, then we copy the term to the human gene. We discard all GO terms assigned to fewer than 100 or to more than 1000 genes. This leaves us with a matrix of 19,111 genes and 1,751 terms.</p>
<h3 id="genomes-project" class="unnumbered">1000 Genomes Project</h3>
<p>We downloaded a filtered (diallelic and 5 or more copies of the minor allele) set of markers from the BEAGLE website and calculated pairwise LD (EUR) for all SNPs in a 1 Mb sliding window:</p>
<p><a href="http://bochet.gcc.biostat.washington.edu/beagle">http://bochet.gcc.biostat.washington.edu/beagle</a></p>
<h2 id="commands">Commands</h2>
<table style="margin:auto" cellpadding="20">
<header><td>
Analysis
</td><td></td><td></td><td>
Figures
</td></header>
<tr><td>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">snpsea</span>
<span class="kw">--snps</span> HDL_Teslovich2010.txt
<span class="kw">--gene-matrix</span> GeneAtlas2004.gct.gz
<span class="kw">--gene-intervals</span> NCBIgenes2013.bed.gz
<span class="kw">--snp-intervals</span> TGP2011.bed.gz
<span class="kw">--null-snps</span> Lango2010.txt.gz
<span class="kw">--score</span> single
<span class="kw">--out</span> <span class="kw">&lt;</span>out<span class="kw">&gt;</span>
<span class="kw">--slop</span> 10e3
<span class="kw">--threads</span> 2
<span class="kw">--null-snpsets</span> 0
<span class="kw">--min-observations</span> 100
<span class="kw">--max-iterations</span> 1e6</code></pre>
</td>
<td></td><td> 
<ul class="incremental">
<li><ul class="incremental">
<li><ul class="incremental">
<li></td>
<td>
Horizontal barplot of p-values:</br></br> <code>snpsea-barplot &lt;out&gt;</code></br></br> Heatmap of conditions and loci:</br></br> <code>snpsea-heatmap &lt;out&gt;</code></br></br> Type 1 error rates for each condition:</br></br> <code>snpsea-type1error &lt;out&gt;</code></br>
<td></tr>
</table>
</li>
</ul></li>
</ul></li>
</ul>
<h1 id="supplementary-figures">Supplementary Figures</h1>
<h2 id="supplementary-figure-1-determining-snp-linkage-intervals">Supplementary Figure 1: Determining SNP linkage intervals</h2>
<div class="figure">
<img src="figures/snp_linkage_intervals.png" />
</div>
<p>We calculated <span class="math">\(r^{2}\)</span> values for all pairs of SNPs within a 1 Mb sliding window along each chromosome. Next, we assigned each of the SNPs from The 1000 Genomes Project Phase I <span class="citation">@consortium_integrated_2012</span> to a linkage interval by identifying each SNP’s furthest upstream and downstream neighbors with <span class="math">\(r^{2}\ge0.5\)</span>. Finally, we extended each interval to recombination hotspots reported by HapMap <span class="citation">@myers_fine-scale_2005</span> with recombination rate &gt;3 cM/Mb.</p>
<h2 id="supplementary-figure-2-counting-genes-in-gwas-snp-linkage-intervals">Supplementary Figure 2: Counting genes in GWAS SNP linkage intervals</h2>
<div class="figure">
<img src="figures/gwas_ngenes.png" />
</div>
<p>A cumulative density plot of the number of genes overlapped by the linkage intervals of GWAS SNPs. We downloaded the GWAS Catalog SNPs on January 17, 2014 and selected the 11,561 SNPs present in the 1000 Genomes Project <span class="citation">@consortium_integrated_2012</span>. Of these SNPs, 2,119 (18%) of them have linkage disequilibrium (LD) intervals that overlap no genes, and 3,756 (32%) overlap a single gene. The remaining 50% of SNPs overlap 2 or more genes. This illustrates the critical issue that many SNPs implicate more than one gene.</p>
<h2 id="supplementary-figure-3-choosing-the-r2-threshold-for-linkage-intervals">Supplementary Figure 3: Choosing the <span class="math">\(r^{2}\)</span> threshold for linkage intervals</h2>
<p>We chose to use <span class="math">\(r^{2}\geq0.5\)</span> due to previous experience <span class="citation">@rossin_proteins_2011</span>. To investigate if this choice influences SNPsea results, we repeated the analysis of 45 red blood cell count-associated SNPs <span class="citation">@van_der_harst_seventy-five_2012</span> using 5 different thresholds (<span class="math">\(r^{2}\ge\)</span> 0.2, 0.4, 0.6, 0.8, 1.0). We also did this for SNPs associated with multiple sclerosis, celiac disease, and HDL cholesterol.</p>
<table><tr><td>
<img src="figures/r2_thresholds-GeneAtlas2004.png" />
</td><td>
<img src="figures/r2_thresholds-GO2013.png" />
</td></tr></table>

<p>Gene Atlas and Gene Ontology (left and right). Each subplot has <span class="math">\(-\text{log}_{10}P\)</span> for <span class="math">\(r^{2}=1\)</span> on the x-axis and <span class="math">\(-\text{log}_{10}P\)</span> on the y-axis for the <span class="math">\(r^{2}\)</span> threshold marked above. Grey lines are significance thresholds after correction testing multiple conditions (cell types, GO annotations). Black points are significant and grey are not. We used the <code>’--score single’</code> option. Red blood cell count SNPs are enriched for <em>hemopoiesis</em> (GO:0030097) (<span class="math">\(P=2\times10^{-5}\)</span>) for linkage intervals with <span class="math">\(r^{2}=(0.6,0.8,1.0)\)</span>. This result falls below the multiple testing threshold at <span class="math">\(r^{2}\ge0.4\)</span>, but remains significant at <span class="math">\(r^{2}\ge0.5\)</span> (see main text).</p>
<h2 id="supplementary-figure-4-each-trait-associated-locus-harbors-a-single-associated-gene">Supplementary Figure 4: Each trait-associated locus harbors a single associated gene</h2>
<div class="figure">
<img src="figures/single_vs_total-RBC-MS-CD-HDL.png" />
</div>
<p>Gene Atlas <span class="citation">@su_gene_2004</span> and Gene Ontology (top and bottom). The x and y axes are<span class="math">\(-\text{log}_{10}P\)</span> for <code>’--score single’ and ’--score total’</code> SNPsea options.</p>
<p>We tested our assumption that a single gene in each trait-associated locus is driving the condition-specificity signal by running SNPsea with the <code>’single’</code> and <code>’total’</code> methods described . We lose power to detect gene expression enrichment when we relax the single gene assumption and instead assume that <em>all</em> genes implicated by LD are driving the signal.</p>
<p>This is in contrast with our observations for GO term enrichment, where p-values appear similar between methods: <code>’single’</code> where we check for the presence of a one or more genes in the locus anotated with a term and <code>’total’</code> where we record the number of genes in the locus annotated with a term.</p>
<h2 id="supplementary-figure-5-type-1-error-estimates">Supplementary Figure 5: Type 1 error estimates</h2>
<div class="figure">
<img src="figures/type1error_GeneAtlas2004.png" />
</div>
<div class="figure">
<img src="figures/type1error_GO2013.png" />
</div>
<p>We sampled 10,000 sets of 100 SNPs uniformly from a list of LD-pruned SNPs <span class="citation">@lango_allen_hundreds_2010</span>. We tested each of the 10,000 sets for enrichment of tissue-specific expression in the Gene Atlas <span class="citation">@su_gene_2004</span> gene expression matrix (top) and for enrichment of annotation with Gene Ontology terms (bottom). For each condition, we show the proportion of the 10,000 enrichment p-values that are below the given thresholds. We observe that the p-values are near the expected values, so the type 1 (false positive) error rate is well-calibrated.</p>
<h1 id="additional-examples">Additional Examples</h1>
<p>We tested SNPsea with the three additional phenotypes listed below with genome-wide significant SNPs <span class="math">\((P\leq5\times10^{-8})\)</span>. When multiple SNPs implicated the same genes, we merged them into a single locus. We tested each phenotype with the Gene Atlas and GO matrices with the <code>’--score single’</code> option. The adjacent heatmaps show Pearson correlation coefficients for all pairs of conditions.</p>
<table style="margin:auto" cellpadding="20">
<tr><td>
Phenotype
</td><td> 
<ul class="incremental">
<li><ul class="incremental">
<li><ul class="incremental">
<li></td><td>
SNPs
</td><td> 
<ul class="incremental">
<li><ul class="incremental">
<li><ul class="incremental">
<li></td><td>
Loci
</td><td> 
<ul class="incremental">
<li><ul class="incremental">
<li><ul class="incremental">
<li></td><td>
Reference
</td></tr>
<tr><td>
Multiple sclerosis
</td><td> 
<ul class="incremental">
<li><ul class="incremental">
<li><ul class="incremental">
<li></td><td>
51
</td><td> 
<ul class="incremental">
<li><ul class="incremental">
<li><ul class="incremental">
<li></td><td>
47
</td><td> 
<ul class="incremental">
<li><ul class="incremental">
<li><ul class="incremental">
<li></td><td>
Supp. Table A, IMSGC WTCCC 2011
</td></tr>
<tr><td>
Celiac disease
</td><td> 
<ul class="incremental">
<li><ul class="incremental">
<li><ul class="incremental">
<li></td><td>
35
</td><td> 
<ul class="incremental">
<li><ul class="incremental">
<li><ul class="incremental">
<li></td><td>
34
</td><td> 
<ul class="incremental">
<li><ul class="incremental">
<li><ul class="incremental">
<li></td><td>
Table 2, Trynka, <em>et al.</em> 2011@trynka_dense_2011
</td></tr>
<tr><td>
HDL cholesterol
</td><td> 
<ul class="incremental">
<li><ul class="incremental">
<li><ul class="incremental">
<li></td><td>
46
</td><td> 
<ul class="incremental">
<li><ul class="incremental">
<li><ul class="incremental">
<li></td><td>
46
</td><td> 
<ul class="incremental">
<li><ul class="incremental">
<li><ul class="incremental">
<li></td><td>
Supp. Table 2, Teslovich, <em>et al.</em> 2010 <span class="citation">@teslovich_biological_2010</span>
</td></tr>
</table>
</li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul></li>
</ul>
<h2 id="supplementary-figure-6-red-blood-cell-count-go-enrichment">Supplementary Figure 6: Red blood cell count GO enrichment</h2>
<div class="figure">
<img src="figures/Red_blood_cell_count-Harst2012-45_SNPs-GO2013-single-pvalues_barplot.png" />
</div>
<p>We observed significant enrichment for <em>hemopoiesis</em> <span class="math">\((2\times10^{-5})\)</span>. The top 25 terms are shown.</p>
<h2 id="supplementary-figure-7-multiple-sclerosis">Supplementary Figure 7: Multiple sclerosis</h2>
<div class="figure">
<img src="figures/Multiple_sclerosis-IMSGC-51_SNPs-GeneAtlas2004-single-pvalues_barplot.png" />
</div>
<p>We observed significant enrichment for 6 cell types. The top 25 of 79 are shown.</p>
<div class="figure">
<img src="figures/Multiple_sclerosis-IMSGC-51_SNPs-GO2013-single-pvalues_barplot.png" />
</div>
<p>We observed significant enrichment for 52 Gene Ontology terms. The top 60 terms are shown.</p>
<h2 id="supplementary-figure-8-celiac-disease">Supplementary Figure 8: Celiac disease</h2>
<div class="figure">
<img src="figures/Celiac_disease-Trynka2011-35_SNPs-GeneAtlas2004-single-pvalues_barplot.png" />
</div>
<p>We observed significant enrichment for 3 cell types. The top 25 of 79 are shown.</p>
<div class="figure">
<img src="figures/Celiac_disease-Trynka2011-35_SNPs-GO2013-single-pvalues_barplot.png" />
</div>
<p>We observed significant enrichment for 28 Gene Ontology terms. The top 40 terms are shown.</p>
<h2 id="supplementary-figure-9-hdl-cholesterol">Supplementary Figure 9: HDL cholesterol</h2>
<div class="figure">
<img src="figures/HDL_cholesterol-Teslovich2010-46_SNPs-GeneAtlas2004-single-pvalues_barplot.png" />
</div>
<p>We observed significant enrichment for liver tissue-specific gene expression. The top 25 of 79 are shown.</p>
<div class="figure">
<img src="figures/HDL_cholesterol-Teslovich2010-46_SNPs-GO2013-single-pvalues_barplot.png" />
</div>
<p>We observed significant enrichment for 13 Gene Ontology terms. The top 25 terms are shown.</p>
    </div>
</div>
</body>
</html>
