

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Algorithm Details &mdash; SNPsea 1.0.3 documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="top" title="SNPsea 1.0.3 documentation" href="index.html"/>
        <link rel="next" title="Installation" href="installation.html"/>
        <link rel="prev" title="Visual Summary" href="visual.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        <a href="index.html" class="fa fa-home"> SNPsea</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
        
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="visual.html">Visual Summary</a><ul>
<li class="toctree-l2"><a class="reference internal" href="visual.html#cartoon">Cartoon</a></li>
<li class="toctree-l2"><a class="reference internal" href="visual.html#flow-chart">Flow Chart</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Algorithm Details</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#step-1-assigning-genes-to-each-snp">Step 1: Assigning genes to each SNP</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-2-calculating-specificity-scores">Step 2: Calculating specificity scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-3-testing-significance">Step 3: Testing significance</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="installation.html#c-libraries">C++ Libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#python-packages">Python Packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#r-packages">R Packages</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="data.html">Data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="data.html#snp-sets">SNP sets</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#gene-matrices">Gene matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#ld-pruned-snps-and-genomic-intervals">LD-pruned SNPs and Genomic Intervals</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="usage.html#options">Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="usage.html#input-file-formats">Input File Formats</a></li>
<li class="toctree-l2"><a class="reference internal" href="usage.html#output-files">Output Files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="output.html">Output Visualizations</a><ul>
<li class="toctree-l2"><a class="reference internal" href="output.html#view-enrichment-of-tissue-specific-gene-expression">View enrichment of tissue-specific gene expression</a></li>
<li class="toctree-l2"><a class="reference internal" href="output.html#view-the-most-specifically-expressed-gene-for-each-snp-tissue-pair">View the most specifically expressed gene for each SNP-tissue pair</a></li>
<li class="toctree-l2"><a class="reference internal" href="output.html#view-the-type-1-error-rate-estimates-for-each-tissue">View the type 1 error rate estimates for each tissue</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="supplementary.html">Supplementary Figures</a><ul>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-1-determining-snp-linkage-intervals">Supplementary Figure 1: Determining SNP linkage intervals</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-2-counting-genes-in-gwas-snp-linkage-intervals">Supplementary Figure 2: Counting genes in GWAS SNP linkage intervals</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-3-choosing-the-threshold-for-linkage-intervals">Supplementary Figure 3: Choosing the <span class="math">\(r^{2}\)</span> threshold for linkage intervals</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-4-each-trait-associated-locus-harbors-a-single-associated-gene">Supplementary Figure 4: Each trait-associated locus harbors a single associated gene</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-5-type-1-error-estimates">Supplementary Figure 5: Type 1 error estimates</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-6-red-blood-cell-count-go-enrichment">Supplementary Figure 6: Red blood cell count GO enrichment</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-7-multiple-sclerosis">Supplementary Figure 7: Multiple sclerosis</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-8-celiac-disease">Supplementary Figure 8: Celiac disease</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-9-hdl-cholesterol">Supplementary Figure 9: HDL cholesterol</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">SNPsea</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Algorithm Details</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/algorithm.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            
  <div class="section" id="algorithm-details">
<h1>Algorithm Details<a class="headerlink" href="#algorithm-details" title="Permalink to this headline">¶</a></h1>
<p>SNPsea tests if genes implicated by risk loci (e.g., those discovered
through genome-wide association (GWA) studies) are specifically
expressed in some conditions over others, and if this specificity is
statistically significant. The program requires two inputs:</p>
<ol class="arabic simple">
<li>A list of SNP identifiers: rs123, 12:456, ...</li>
<li>A matrix of genes and conditions, such as:<ul>
<li>Gene expression profiles of multiple different cell types.</li>
<li>Ontology terms and presence/absence 1/0 values for each gene in
each term.</li>
</ul>
</li>
</ol>
<p>For example, SNPsea can be used to find tissues or cell types whose
function is likely to be influenced by genes in risk loci. If the genes
in risk loci are used in relatively few cell types, we hypothesize that
they are likely to affect those cell types’ unique functions. This
assumes that expression specificity is a good indicator of a gene’s
importance to the unique function of the cell type.</p>
<p>For a given set of SNPs associated to some phenotype, SNPsea tests
whether all implicated genes, in aggregate, are enriched for specificity
to a condition in a user-provided matrix of genes and
conditions/annotations. The algorithm consists of three steps:</p>
<ul class="simple">
<li><strong>Step 1: Assigning genes to each SNP</strong><ul>
<li>We use linkage disequilibrium (LD) to identify the genes
implicated by each SNP.</li>
</ul>
</li>
<li><strong>Step 2: Calculating specificity scores</strong><ul>
<li>We look up implicated genes in a user-provided matrix and
calculate a specificity score for each annotation/condition based
on the values of these genes.</li>
</ul>
</li>
<li><strong>Step 3: Testing significance</strong><ul>
<li>We compare the specificity scores to a null distribution of scores
obtained with random sets of matched SNP sets and compute an
empirical <span class="math">\(P\)</span>-value.</li>
</ul>
</li>
</ul>
<div class="section" id="step-1-assigning-genes-to-each-snp">
<h2>Step 1: Assigning genes to each SNP<a class="headerlink" href="#step-1-assigning-genes-to-each-snp" title="Permalink to this headline">¶</a></h2>
<p>Accurate analyses must address the critical issue that SNPs frequently
implicate a region with multiple different genes (<strong>Supplementary Figure
2</strong>). The challenge is to find evidence to show which of those genes are
associated with a given trait.</p>
<p>We determine the genes plausibly implicated by each trait-associated SNP
using a previously described strategy (<strong>Supplementary Figure 1</strong> and
Rossin <em>et al.</em> 2011). First, we define the linkage interval for a given
SNP as the span between the furthest correlated SNPs <span class="math">\(r^{2}&gt;0.5\)</span>
(EUR) within a 1 Mb window (1000 Genomes Consortium 2012). Next, we
extend the interval to the nearest recombination hotspots with
recombination rate &gt;3 cM/Mb (Myers <em>et al.</em> 2005). To address the case
when no genes overlap an interval, we provide an option for SNPsea to
extend the interval up- and downstream (by default 10 Kb).</p>
<p>Most frequently, we find multiple genes <span class="math">\((m_{k}&gt;1)\)</span> in a single
SNP locus <span class="math">\(k\)</span>. We expect many loci with multiple genes because of
regions with high LD across long stretches of a chromosome. Less
frequently, a locus has a single gene <span class="math">\((m_{k}=1)\)</span>, and loci with
no genes <span class="math">\((m_{k}=0)\)</span> are discarded.</p>
<p>After each SNP has been assigned an interval and a set of genes
overlapping the interval, we merge SNPs with shared genes into a single
locus to avoid multiple-counting of genes.</p>
<div class="section" id="two-score-options">
<h3>Two score options<a class="headerlink" href="#two-score-options" title="Permalink to this headline">¶</a></h3>
<p>By default, SNPsea assumes one gene in each associated locus is
associated with the given trait. We also include the option to assume
all genes within a locus are associated. We compare results of the two
options with four phenotypes (<strong>Supplementary Figure 4</strong>).</p>
<ol class="arabic simple">
<li>The <tt class="docutils literal"><span class="pre">’--score</span> <span class="pre">single’</span></tt> method (default option) assumes that a
single gene in each locus is associated with the given phenotype. For
each condition, we choose the gene in each locus with the greatest
specificity to that condition.</li>
<li>The <tt class="docutils literal"><span class="pre">’--score</span> <span class="pre">total’</span></tt> method assumes that all genes in a SNP’s
linkage interval are associated. We account for all linked genes when
calculating scores.</li>
</ol>
</div>
</div>
<div class="section" id="step-2-calculating-specificity-scores">
<h2>Step 2: Calculating specificity scores<a class="headerlink" href="#step-2-calculating-specificity-scores" title="Permalink to this headline">¶</a></h2>
<p>SNPsea uses different algorithms for matrices with continuous or binary
values. Before running SNPsea, a matrix with continuous values must be
normalized so that columns are directly comparable. <em>It is not
appropriate to use this method on a “raw” matrix of expression values</em>.</p>
<div class="section" id="specificity-for-a-matrix-of-continuous-values">
<h3>Specificity for a matrix of continuous values<a class="headerlink" href="#specificity-for-a-matrix-of-continuous-values" title="Permalink to this headline">¶</a></h3>
<p>We extend an approach we have previoulsy described in detail (Hu <em>et
al.</em> 2011). Let <span class="math">\(A\)</span> denote a continuous gene expression matrix
with <span class="math">\(m\)</span> genes and <span class="math">\(n\)</span> conditions. First, we normalize the
expression of each gene by dividing each value by the L2 norm of the
genes values in different conditions.</p>
<div class="math">
\[A'_{i,j}=\dfrac{A_{i,j}}{\sqrt{A_{i,1}^{2}+A_{i,2}^{2}+\cdots+A_{i,n}^{2}}}\]</div>
<p>The resulting matrix <span class="math">\(A'\)</span> has values <span class="math">\(A'_{i,j}\)</span> between 0
and 1 indicating specificity of gene <span class="math">\(i\)</span> to condition <span class="math">\(j\)</span>. A
value <span class="math">\(A'_{i,j}=1\)</span> indicates that gene <span class="math">\(i\)</span> is exclusively
expressed in condition <span class="math">\(j\)</span>, and <span class="math">\(A'_{i,j}=0\)</span> indicates that
gene <span class="math">\(i\)</span> is not expressed in condition <span class="math">\(j\)</span>.</p>
<p>Next, we transform <span class="math">\(A'\)</span> to a matrix <span class="math">\(A''\)</span> of non-parametric
condition-specificity percentiles as follows. For each condition
<span class="math">\(j\)</span>, we rank the values of <span class="math">\(A'_{,j}\)</span> in ascending order and
divide them by the number of genes <span class="math">\(m\)</span>, resulting in percentiles
between 0 and 1 where a lower value indicates greater specificity to the
given condition.</p>
<div class="math">
\[A''_{i,j}=\dfrac{\text{Rank}_{j}(A'_{i,j})}{m}\]</div>
</div>
<div class="section" id="locus-scores-for-a-matrix-of-continuous-values">
<h3>Locus scores for a matrix of continuous values<a class="headerlink" href="#locus-scores-for-a-matrix-of-continuous-values" title="Permalink to this headline">¶</a></h3>
<p>We create a new matrix <span class="math">\(P\)</span>, where each value <span class="math">\(P_{k,j}\)</span> is a
score for a SNP locus <span class="math">\(k\)</span> and a condition <span class="math">\(j\)</span>. The locus
scores <span class="math">\(P_{,j}\)</span> for a single condition <span class="math">\(j\)</span> are approximately
uniformly distributed for a set of randomly selected loci under the
following assumption: for the set of genes in a given SNP locus
<span class="math">\(I_{k}\)</span>, the values <span class="math">\(A''_{i\in I_{k},j}\)</span> are random,
independent, and approximately uniformly distributed. We’ll come back to
this assumption later when testing significance in Step 3 below.</p>
</div>
<div class="section" id="score-single-default">
<h3><tt class="docutils literal"><span class="pre">’--score</span> <span class="pre">single’</span> <span class="pre">(default)</span></tt><a class="headerlink" href="#score-single-default" title="Permalink to this headline">¶</a></h3>
<p>This approach assumes one gene in each SNP locus is associated with the
trait.</p>
<p>For each locus-condition pair <span class="math">\((k,j)\)</span>, we choose the single gene
<span class="math">\(i\)</span> in locus <span class="math">\(k\)</span> with greatest specificity to condition
<span class="math">\(j\)</span> among the <span class="math">\(m_{k}\)</span> genes in the locus, as previously
described in Hu <em>et al.</em> (Hu <em>et al.</em> 2011). Let <span class="math">\(g_{k}\)</span> denote
this most specific gene, so that
<span class="math">\(A''_{g_{k},j}=\text{Min}_{i\in I_{k}}(A''_{i,j})\)</span> where
<span class="math">\(I_{k}\)</span> denotes the set of genes in locus <span class="math">\(k\)</span>. If we assume
values of <span class="math">\(A''_{i\in I_{k},j}\)</span> are uniformly distributed for a
given condition <span class="math">\(j\)</span> and genes <span class="math">\(i\in I_{k}\)</span>, then the
probability of obtaining a value equal to or less than
<span class="math">\(A''_{g_{k},j}\)</span> is as follows:</p>
<div class="math">
\[P_{k,j}=1-(1-\text{Min}_{i\in I_{k}}(A''_{i,j}))^{m_{k}}\]</div>
</div>
<div class="section" id="score-total">
<h3><tt class="docutils literal"><span class="pre">’--score</span> <span class="pre">total’</span></tt><a class="headerlink" href="#score-total" title="Permalink to this headline">¶</a></h3>
<p>This assumes all genes in a given SNP locus are associated with a trait
— in most circumstances, we consider this model to be unlikely in most
situations. We compute the probability of observing values
<span class="math">\(A''_{i\in I_{k}}\)</span> for some locus <span class="math">\(k\)</span> as the product of
percentiles. This assumes <span class="math">\(A''_{i\in I_{k}}\)</span> values are uniformly
distributed.</p>
<div class="math">
\[\begin{split}\begin{aligned}
P_{k,j} &amp; = &amp; \int_{x}^{\infty}\Gamma(m_{k},1)\ \ \text{for}\ \ x=\sum_{i\in I_{k}}-\text{ln}A''_{i,j}\end{aligned}\end{split}\]</div>
</div>
<div class="section" id="locus-scores-for-a-matrix-of-binary-values">
<h3>Locus scores for a matrix of binary values<a class="headerlink" href="#locus-scores-for-a-matrix-of-binary-values" title="Permalink to this headline">¶</a></h3>
<p>Let <span class="math">\(B\)</span> denote a binary matrix (1=present, 0=absent) with
<span class="math">\(m\)</span> genes and <span class="math">\(n\)</span> conditions. Let <span class="math">\(m_{j}\)</span> denote the
number of genes present in condition <span class="math">\(j\)</span>. Let <span class="math">\(m_{k}\)</span> denote
the number of genes in locus <span class="math">\(k\)</span> and <span class="math">\(m_{k,j}\le m_{k}\)</span>
denote the number of genes in locus <span class="math">\(k\)</span> that are present in
condition <span class="math">\(j\)</span>.</p>
<p>We provide two options to calculate locus scores. By default, we account
for presence or absence of any of the <span class="math">\(m_{k}\)</span> genes in condition
<span class="math">\(j\)</span>, as shown below (<tt class="docutils literal"><span class="pre">’--score</span> <span class="pre">single’</span></tt>). Alternatively, we
account for the number of genes in a given locus (<tt class="docutils literal"><span class="pre">’--score</span> <span class="pre">total’</span></tt>).</p>
<table border="1" class="docutils">
<colgroup>
<col width="44%" />
<col width="56%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">`’&#8211;score single’</th>
<th class="head"><tt class="docutils literal"><span class="pre">’--score</span> <span class="pre">total’</span></tt></th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><span class="math">\(P_{k,j}=\begin{cases} 1-p(0) &amp; m_{k,j}&gt;0\\ 1 &amp; m_{k,j}=0 \end{cases}\)</span></td>
<td><span class="math">\(P_{k,j}=\begin{cases} 1-\sum_{x=0}^{m_{k,j}-1}p(x) &amp; m_{k,j}&gt;0\\ 1 &amp; m_{k,j}=0 \end{cases}\)</span></td>
</tr>
</tbody>
</table>
<p>where</p>
<div class="math">
\[p(x)=\dfrac{{m_{j} \choose x}{m-m_{j} \choose m_{k}-x}}{{m \choose m_{k}}}\]</div>
</div>
<div class="section" id="condition-specificity-scores">
<h3>Condition specificity scores<a class="headerlink" href="#condition-specificity-scores" title="Permalink to this headline">¶</a></h3>
<p>For both continuous and binary matrices, we define a specificity score
<span class="math">\(S_{j}\)</span> for each condition <span class="math">\(j\)</span> as the aggregate of
<span class="math">\(P_{k,j}\)</span> values across SNP loci:</p>
<div class="math">
\[S_{j}=\sum_{k}-\text{log}P_{k,j}\]</div>
</div>
</div>
<div class="section" id="step-3-testing-significance">
<h2>Step 3: Testing significance<a class="headerlink" href="#step-3-testing-significance" title="Permalink to this headline">¶</a></h2>
<div class="section" id="analytical-p-values">
<h3>Analytical p-values<a class="headerlink" href="#analytical-p-values" title="Permalink to this headline">¶</a></h3>
<p>We previously found that aggregating the <span class="math">\(P_{k,j}\)</span> scores and
determining a <span class="math">\(P\)</span>-value analytically from a distribution results
in inaccurate p-values (Hu <em>et al.</em> 2011). <span class="math">\(A''_{i,j}\)</span> values may
be relatively uniform genome-wide, but proximate genes often have shared
functions. The genome has a complex correlation structure of linkage
disequilibrium, gene density, gene size and function that is challenging
to model analytically. We use the sampling strategy described below
instead.</p>
</div>
<div class="section" id="permutation-p-values">
<h3>Permutation p-values<a class="headerlink" href="#permutation-p-values" title="Permalink to this headline">¶</a></h3>
<p>For each condition, we use a sampling approach to calculate an empirical
p-value. This is the tail probability of observing a
condition-specificity score greater or equal to <span class="math">\(S_{j}\)</span>. We obtain
the distribution empirically with null SNP sets.</p>
<p>We compute specificity scores <span class="math">\(S\)</span> for random SNP sets. Each SNP in
a null set is matched to a SNP in the user’s set on the number of linked
genes. To adequately sample genes from the entire genome, we sample SNP
sets from a list of LD-pruned SNPs (subset of SNPs in 1000 Genomes
Project) (Lango Allen <em>et al.</em> 2010).</p>
<p>For each condition <span class="math">\(j\)</span>, we calculate an exact permutation p-value
(Phipson <em>et al.</em> 2010). Let <span class="math">\(a_{j}\)</span> denote the number of sampled
SNP sets (e.g. 10,000) and let <span class="math">\(b_{j}\)</span> denote how many null
specificity scores are greater than or equal to the user’s score
<span class="math">\(S_{j}\)</span>:</p>
<div class="math">
\[p_{j}=\dfrac{b_{j}+1}{a_{j}+1}\]</div>
<p>We implemented adaptive sampling to calculate p-values efficiently. As
each condition is tested for significance, we increase the number of
iterations to resolve significant p-values and save computation by using
fewer iterations for less significant p-values. Two options allow the
user to control the adaptive sampling:</p>
<ol class="arabic simple">
<li><tt class="docutils literal"><span class="pre">’--max-iterations</span> <span class="pre">N’</span></tt> The maximum number of iterations for each
condition. We stop testing a condition after sampling <span class="math">\(N\)</span> SNP
sets.</li>
<li><tt class="docutils literal"><span class="pre">’--min-observations</span> <span class="pre">N’</span></tt> The minimum number of observed null
specificity scores greater than or equal to <span class="math">\(S_{j}\)</span> required to
stop sampling SNP sets for a condition <span class="math">\(j\)</span>.</li>
</ol>
</div>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="installation.html" class="btn btn-neutral float-right" title="Installation"/>Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="visual.html" class="btn btn-neutral" title="Visual Summary"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014, Kamil Slowikowski.
    </p>
  </div>

  <a href="https://github.com/snide/sphinx_rtd_theme">Sphinx theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.3',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>