

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Data &mdash; SNPsea 1.0.3 documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="top" title="SNPsea 1.0.3 documentation" href="index.html"/>
        <link rel="next" title="Usage" href="usage.html"/>
        <link rel="prev" title="Installation" href="installation.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        <a href="index.html" class="fa fa-home"> SNPsea</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
        
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="visual.html">Visual Summary</a><ul>
<li class="toctree-l2"><a class="reference internal" href="visual.html#cartoon">Cartoon</a></li>
<li class="toctree-l2"><a class="reference internal" href="visual.html#flow-chart">Flow Chart</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="algorithm.html">Algorithm Details</a><ul>
<li class="toctree-l2"><a class="reference internal" href="algorithm.html#step-1-assigning-genes-to-each-snp">Step 1: Assigning genes to each SNP</a></li>
<li class="toctree-l2"><a class="reference internal" href="algorithm.html#step-2-calculating-specificity-scores">Step 2: Calculating specificity scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="algorithm.html#step-3-testing-significance">Step 3: Testing significance</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="installation.html#c-libraries">C++ Libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#python-packages">Python Packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#r-packages">R Packages</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#snp-sets">SNP sets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#gene-matrices">Gene matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ld-pruned-snps-and-genomic-intervals">LD-pruned SNPs and Genomic Intervals</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="usage.html#options">Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="usage.html#input-file-formats">Input File Formats</a></li>
<li class="toctree-l2"><a class="reference internal" href="usage.html#output-files">Output Files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="output.html">Output Visualizations</a><ul>
<li class="toctree-l2"><a class="reference internal" href="output.html#view-enrichment-of-tissue-specific-gene-expression">View enrichment of tissue-specific gene expression</a></li>
<li class="toctree-l2"><a class="reference internal" href="output.html#view-the-most-specifically-expressed-gene-for-each-snp-tissue-pair">View the most specifically expressed gene for each SNP-tissue pair</a></li>
<li class="toctree-l2"><a class="reference internal" href="output.html#view-the-type-1-error-rate-estimates-for-each-tissue">View the type 1 error rate estimates for each tissue</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="supplementary.html">Supplementary Figures</a><ul>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-1-determining-snp-linkage-intervals">Supplementary Figure 1: Determining SNP linkage intervals</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-2-counting-genes-in-gwas-snp-linkage-intervals">Supplementary Figure 2: Counting genes in GWAS SNP linkage intervals</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-3-choosing-the-threshold-for-linkage-intervals">Supplementary Figure 3: Choosing the <span class="math">\(r^{2}\)</span> threshold for linkage intervals</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-4-each-trait-associated-locus-harbors-a-single-associated-gene">Supplementary Figure 4: Each trait-associated locus harbors a single associated gene</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-5-type-1-error-estimates">Supplementary Figure 5: Type 1 error estimates</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-6-red-blood-cell-count-go-enrichment">Supplementary Figure 6: Red blood cell count GO enrichment</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-7-multiple-sclerosis">Supplementary Figure 7: Multiple sclerosis</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-8-celiac-disease">Supplementary Figure 8: Celiac disease</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-9-hdl-cholesterol">Supplementary Figure 9: HDL cholesterol</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">SNPsea</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Data</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/data.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            
  <div class="section" id="data">
<h1>Data<a class="headerlink" href="#data" title="Permalink to this headline">¶</a></h1>
<div class="code bash highlight-python"><div class="highlight"><pre>cd snpsea
curl -LOk http://files.figshare.com/1504037/SNPsea_data_20140520.zip
unzip SNPsea_data_20140520.zip
</pre></div>
</div>
<p>Download the compressed archive with data required to perform this
analysis (138M). The direct link to the zip shown above may be out of
date and fail to load. If so, please visit the link below instead:</p>
<p><a class="reference external" href="http://dx.doi.org/10.6084/m9.figshare.871430">http://dx.doi.org/10.6084/m9.figshare.871430</a></p>
<p>Contents of the compressed archive with data:</p>
<div class="highlight-python"><div class="highlight"><pre>Celiac_disease-Trynka2011-35_SNPs.gwas
HDL_cholesterol-Teslovich2010-46_SNPs.gwas
Multiple_sclerosis-IMSGC-51_SNPs.gwas
Red_blood_cell_count-Harst2012-45_SNPs.gwas

GeneAtlas2004.gct.gz  # Gene Atlas 2004 Affymetrix expression matrix
ImmGen2012.gct.gz     # ImmGen 2012 Affymetrix expression matrix
FANTOM2014.gct.gz     # FANTOM5 2014 CAGE matrix
GO2013.gct.gz         # Gene Ontology 2013 binary annotation matrix

NCBIgenes2013.bed.gz  # NCBI gene intervals
Lango2010.txt.gz      # LD-pruned SNPs
TGP2011.bed.gz        # 1000 Genomes Project SNP linkage intervals
</pre></div>
</div>
<div class="section" id="snp-sets">
<h2>SNP sets<a class="headerlink" href="#snp-sets" title="Permalink to this headline">¶</a></h2>
<table border="1" class="docutils">
<colgroup>
<col width="18%" />
<col width="6%" />
<col width="6%" />
<col width="70%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Phenotype</th>
<th class="head">SNPs</th>
<th class="head">Loci</th>
<th class="head">Reference</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Celiac disease</td>
<td>35</td>
<td>34</td>
<td>Table 2 (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/22057235">Trynka, et al. 2011</a>)</td>
</tr>
<tr class="row-odd"><td>HDL cholesterol</td>
<td>46</td>
<td>46</td>
<td>Supp. Table 2 (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/20686565">Teslovich, et al. 2010</a>)</td>
</tr>
<tr class="row-even"><td>Multiple sclerosis</td>
<td>51</td>
<td>47</td>
<td>Supp. Table A (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/21833088">IMSGC WTCCC 2011</a>)</td>
</tr>
<tr class="row-odd"><td>Red blood cell count</td>
<td>45</td>
<td>45</td>
<td>Table 1 (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/23222517">Harst et al. 2012</a>)</td>
</tr>
</tbody>
</table>
<div class="section" id="celiac-disease-trynka2011-35-snps-gwas">
<h3>Celiac_disease-Trynka2011-35_SNPs.gwas<a class="headerlink" href="#celiac-disease-trynka2011-35-snps-gwas" title="Permalink to this headline">¶</a></h3>
<p>35 SNPs associated with Celiac disease taken from Table 2. Positions are
on hg19. All SNPs have <span class="math">\(P \le 5e-8\)</span>.</p>
<blockquote>
<div>Trynka G, Hunt KA, Bockett NA, et al. Dense genotyping identifies
and localizes multiple common and rare variant association signals
in celiac disease. Nat Genet. 2011;43(12):1193-201.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/22057235">Pubmed</a></div></blockquote>
</div>
<div class="section" id="hdl-cholesterol-teslovich2010-46-snps-gwas">
<h3>HDL_cholesterol-Teslovich2010-46_SNPs.gwas<a class="headerlink" href="#hdl-cholesterol-teslovich2010-46-snps-gwas" title="Permalink to this headline">¶</a></h3>
<p>46 SNPs associated with HDL taken from Supplementary Table 2. Positions
are on hg19. All SNPs have <span class="math">\(P \le 5e-8\)</span>.</p>
<blockquote>
<div>Teslovich TM, Musunuru K, Smith AV, et al. Biological, clinical and
population relevance of 95 loci for blood lipids. Nature.
2010;466(7307):707-13.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/20686565">Pubmed</a></div></blockquote>
</div>
<div class="section" id="multiple-sclerosis-imsgc-51-snps-gwas">
<h3>Multiple_sclerosis-IMSGC-51_SNPs.gwas<a class="headerlink" href="#multiple-sclerosis-imsgc-51-snps-gwas" title="Permalink to this headline">¶</a></h3>
<p>51 SNPs associated with Multiple Sclerosis taken from Supplementary
Table A. Positions are on hg19. All SNPs have <span class="math">\(P \le 5e-8\)</span>.</p>
<blockquote>
<div>Sawcer S, Hellenthal G, Pirinen M, et al. Genetic risk and a primary
role for cell-mediated immune mechanisms in multiple sclerosis.
Nature. 2011;476(7359):214-9.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/21833088">Pubmed</a></div></blockquote>
</div>
<div class="section" id="red-blood-cell-count-harst2012-45-snps-gwas">
<h3>Red_blood_cell_count-Harst2012-45_SNPs.gwas<a class="headerlink" href="#red-blood-cell-count-harst2012-45-snps-gwas" title="Permalink to this headline">¶</a></h3>
<p>45 SNPs associated with red blood cell count (RBC) taken from Table 1.
Positions are on hg19. All SNPs have <span class="math">\(P \le 5e-8\)</span>.</p>
<blockquote>
<div>van der Harst P, Zhang W, Mateo leach I, et al. Seventy-five genetic
loci influencing the human red blood cell. Nature.
2012;492(7429):369-75.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/23222517">Pubmed</a></div></blockquote>
</div>
</div>
<div class="section" id="gene-matrices">
<h2>Gene matrices<a class="headerlink" href="#gene-matrices" title="Permalink to this headline">¶</a></h2>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="5%" />
<col width="8%" />
<col width="17%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Type</th>
<th class="head">Genes</th>
<th class="head">Conditions</th>
<th class="head">Species</th>
<th class="head">Reference</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Affy</td>
<td>17581</td>
<td>79 tissues</td>
<td>homo sapiens</td>
<td><a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE1133">GeneAtlas 2004</a></td>
</tr>
<tr class="row-odd"><td>Affy</td>
<td>15139</td>
<td>249 cells</td>
<td>mus musculus</td>
<td><a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE15907">ImmGen 2012</a></td>
</tr>
<tr class="row-even"><td>CAGE</td>
<td>18502</td>
<td>533 cells</td>
<td>homo sapiens</td>
<td><a class="reference external" href="http://fantom.gsc.riken.jp/5/data/">FANTOM5 2014</a></td>
</tr>
<tr class="row-odd"><td>Binary</td>
<td>19111</td>
<td>1751 terms</td>
<td>homo sapiens, mus musculus</td>
<td><a class="reference external" href="http://www.geneontology.org">Gene Ontology</a> 2013, <a class="reference external" href="http://www.ncbi.nlm.nih.gov/homologene">Homologene</a></td>
</tr>
</tbody>
</table>
<div class="section" id="geneatlas2004-gct-gz">
<h3>GeneAtlas2004.gct.gz<a class="headerlink" href="#geneatlas2004-gct-gz" title="Permalink to this headline">¶</a></h3>
<p>Gene expression data for 79 human tissues from
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE1133">GSE1133</a>.
We averaged the expression values for tissue replicates. For each gene,
we selected the single probe with the largest minimum value. Finally, we
converted the file to
<a class="reference external" href="http://www.broadinstitute.org/cancer/software/genepattern/gp_guides/file-formats/sections/gct">GCT</a>
format.</p>
<blockquote>
<div>Su AI et al. A gene atlas of the mouse and human protein-encoding
transcriptomes. Proc Natl Acad Sci U S A, 2004 Apr 9;101(16):6062-7.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE1133">PubMed</a></div></blockquote>
</div>
<div class="section" id="go2013-gct-gz">
<h3>GO2013.gct.gz<a class="headerlink" href="#go2013-gct-gz" title="Permalink to this headline">¶</a></h3>
<p>A
<a class="reference external" href="http://www.broadinstitute.org/cancer/software/genepattern/gp_guides/file-formats/sections/gct">GCT</a>
formatted gene matrix with 1,751 annotation terms (1s and 0s indicating
presence or absence of the gene in a Gene Ontology term).</p>
<p>We downloaded the OBO file from <a class="reference external" href="http://www.geneontology.org">Gene
Ontology</a> (data-version: 2013-06-29, CVS
revision: 9700).</p>
<p>For each gene, we climbed the hierarchy of ontology terms and applied
parental terms. If a gene is annotated with some term <span class="math">\(T\)</span>, we also
add all of the terms that are parents of <span class="math">\(T\)</span>. We copy terms
between homologous genes using
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/homologene">Homologene</a> data. If a mouse
gene is annotated with some term and the human homolog is not, then we
copy the term to the human gene. We discard all GO terms assigned to
fewer than 100 or to more than 1000 genes. This leaves us with a matrix
of 19,111 genes and 1,751 terms.</p>
</div>
<div class="section" id="immgen2012-gct-gz">
<h3>ImmGen2012.gct.gz<a class="headerlink" href="#immgen2012-gct-gz" title="Permalink to this headline">¶</a></h3>
<p>Gene expression data for 249 blood cell types from
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE15907">GSE15907</a>.
We averaged cell type replicates. For each gene, we selected the single
probe with the largest minimum.</p>
</div>
<div class="section" id="fantom2014-gct-gz">
<h3>FANTOM2014.gct.gz<a class="headerlink" href="#fantom2014-gct-gz" title="Permalink to this headline">¶</a></h3>
<p>CAGE data for 533 human cell types from
<a class="reference external" href="http://fantom.gsc.riken.jp/5/data/">FANTOM5</a>. We averaged cell type
replicates. We discarded CAGE entries with 0 or multiple corresponding
NCBI Entrez IDs. Then, we summed the CAGE entries for each gene.</p>
</div>
</div>
<div class="section" id="ld-pruned-snps-and-genomic-intervals">
<h2>LD-pruned SNPs and Genomic Intervals<a class="headerlink" href="#ld-pruned-snps-and-genomic-intervals" title="Permalink to this headline">¶</a></h2>
<div class="section" id="lango2010-txt-gz">
<h3>Lango2010.txt.gz<a class="headerlink" href="#lango2010-txt-gz" title="Permalink to this headline">¶</a></h3>
<p>A list of SNPs that span the whole genome, pruned by linkage
disequilibrium (LD). SNPsea samples null SNP sets matched on the number
of genes in the user&#8217;s SNP set from this list. See this paper for more
information:</p>
<blockquote>
<div>Lango allen H, Estrada K, Lettre G, et al. Hundreds of variants
clustered in genomic loci and biological pathways affect human
height. Nature. 2010;467(7317):832-8.
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/20881960">Pubmed</a></div></blockquote>
</div>
<div class="section" id="ncbigenes2013-bed-gz">
<h3>NCBIgenes2013.bed.gz<a class="headerlink" href="#ncbigenes2013-bed-gz" title="Permalink to this headline">¶</a></h3>
<p>All human start and stop positions taken from:</p>
<blockquote>
<div><a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene2refseq.gz">ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene2refseq.gz</a></div></blockquote>
</div>
<div class="section" id="tgp2011-bed-gz">
<h3>TGP2011.bed.gz<a class="headerlink" href="#tgp2011-bed-gz" title="Permalink to this headline">¶</a></h3>
<p>Linkage intervals for a filtered set of SNPs from the <a class="reference external" href="http://www.1000genomes.org/">1000 Genomes
Project</a> Phase 1 (May 21, 2011). We
downloaded a filtered (diallelic and 5 or more copies of the minor
allele) set of markers from the
<a class="reference external" href="http://bochet.gcc.biostat.washington.edu/beagle/1000_Genomes.phase1_release_v3">BEAGLE</a>
website and calculated pairwise LD (EUR) for all SNPs in a 1 Mb sliding
window. The linkage intervals were extended to the nearest
<a class="reference external" href="http://hapmap.ncbi.nlm.nih.gov/downloads/">HapMap</a> recombination
hotspot with &gt;3 cM/Mb recombination rate.</p>
</div>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="usage.html" class="btn btn-neutral float-right" title="Usage"/>Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral" title="Installation"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014, Kamil Slowikowski.
    </p>
  </div>

  <a href="https://github.com/snide/sphinx_rtd_theme">Sphinx theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.3',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>