

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Installation &mdash; SNPsea 1.0.3 documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="top" title="SNPsea 1.0.3 documentation" href="index.html"/>
        <link rel="next" title="Data" href="data.html"/>
        <link rel="prev" title="Algorithm Details" href="algorithm.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        <a href="index.html" class="fa fa-home"> SNPsea</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
        
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="visual.html">Visual Summary</a><ul>
<li class="toctree-l2"><a class="reference internal" href="visual.html#cartoon">Cartoon</a></li>
<li class="toctree-l2"><a class="reference internal" href="visual.html#flow-chart">Flow Chart</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="algorithm.html">Algorithm Details</a><ul>
<li class="toctree-l2"><a class="reference internal" href="algorithm.html#step-1-assigning-genes-to-each-snp">Step 1: Assigning genes to each SNP</a></li>
<li class="toctree-l2"><a class="reference internal" href="algorithm.html#step-2-calculating-specificity-scores">Step 2: Calculating specificity scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="algorithm.html#step-3-testing-significance">Step 3: Testing significance</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#c-libraries">C++ Libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="#python-packages">Python Packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="#r-packages">R Packages</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="data.html">Data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="data.html#snp-sets">SNP sets</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#gene-matrices">Gene matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#ld-pruned-snps-and-genomic-intervals">LD-pruned SNPs and Genomic Intervals</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="usage.html#options">Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="usage.html#input-file-formats">Input File Formats</a></li>
<li class="toctree-l2"><a class="reference internal" href="usage.html#output-files">Output Files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="output.html">Output Visualizations</a><ul>
<li class="toctree-l2"><a class="reference internal" href="output.html#view-enrichment-of-tissue-specific-gene-expression">View enrichment of tissue-specific gene expression</a></li>
<li class="toctree-l2"><a class="reference internal" href="output.html#view-the-most-specifically-expressed-gene-for-each-snp-tissue-pair">View the most specifically expressed gene for each SNP-tissue pair</a></li>
<li class="toctree-l2"><a class="reference internal" href="output.html#view-the-type-1-error-rate-estimates-for-each-tissue">View the type 1 error rate estimates for each tissue</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="supplementary.html">Supplementary Figures</a><ul>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-1-determining-snp-linkage-intervals">Supplementary Figure 1: Determining SNP linkage intervals</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-2-counting-genes-in-gwas-snp-linkage-intervals">Supplementary Figure 2: Counting genes in GWAS SNP linkage intervals</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-3-choosing-the-threshold-for-linkage-intervals">Supplementary Figure 3: Choosing the <span class="math">\(r^{2}\)</span> threshold for linkage intervals</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-4-each-trait-associated-locus-harbors-a-single-associated-gene">Supplementary Figure 4: Each trait-associated locus harbors a single associated gene</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-5-type-1-error-estimates">Supplementary Figure 5: Type 1 error estimates</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-6-red-blood-cell-count-go-enrichment">Supplementary Figure 6: Red blood cell count GO enrichment</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-7-multiple-sclerosis">Supplementary Figure 7: Multiple sclerosis</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-8-celiac-disease">Supplementary Figure 8: Celiac disease</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-9-hdl-cholesterol">Supplementary Figure 9: HDL cholesterol</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">SNPsea</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Installation</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/installation.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            
  <div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p><strong>On Linux 64-bit, you may use the provided executable</strong></p>
<p>This runs on kernel 2.6.18 and newer:
<a class="reference external" href="https://github.com/slowkow/snpsea/releases">https://github.com/slowkow/snpsea/releases</a></p>
<p><strong>Otherwise, you must build the executable from source</strong></p>
<p>The source code is available: <a class="reference external" href="https://github.com/slowkow/snpsea">https://github.com/slowkow/snpsea</a></p>
<p><strong>Mac:</strong> To compile C++ code with the required dependencies, you need
XCode and MacPorts: <a class="reference external" href="http://guide.macports.org/#installing.xcode">http://guide.macports.org/#installing.xcode</a></p>
<p>Install the dependencies:</p>
<div class="code bash highlight-python"><div class="highlight"><pre># Ubuntu
sudo apt-get install build-essential libopenmpi-dev libgsl0-dev

# Mac
#   First, install port (MacPorts): http://www.macports.org/
#   Next, use it to install the dependencies:
sudo port selfupdate &amp;&amp; sudo port install gcc48 openmpi gsl

# Broad Institute
#   Add this line to ~/.my.bashrc or ~/.my.cshrc
use .gcc-4.8.1 .openmpi-1.4 .gsl-1.14
</pre></div>
</div>
<p>Download and compile the code:</p>
<div class="code bash highlight-python"><div class="highlight"><pre>#   Clone with git; easily get updates with &#39;git pull&#39;:
git clone https://github.com/slowkow/snpsea.git
cd snpsea

#   If you don&#39;t have git:
curl -LOk https://github.com/slowkow/snpsea/archive/master.zip
unzip master.zip; cd snpsea-master

cd src; make               #   Compile.
cp ../bin/snpsea* ~/bin/   #   Copy the executables wherever you like.
</pre></div>
</div>
<div class="section" id="c-libraries">
<h2>C++ Libraries<a class="headerlink" href="#c-libraries" title="Permalink to this headline">¶</a></h2>
<p>To compile SNPsea, you will need a modern C++ compiler that supports
<a class="reference external" href="https://gcc.gnu.org/projects/cxx0x.html">c++0x</a> and the dependencies
listed below. I compiled successfully with gcc versions 4.6.3 (the default
version for Ubuntu 12.04) and 4.8.1.</p>
<p><a class="reference external" href="https://github.com/slowkow/intervaltree">intervaltree</a></p>
<blockquote>
<div>a minimal C++ interval tree implementation</div></blockquote>
<p><a class="reference external" href="http://eigen.tuxfamily.org">Eigen</a></p>
<blockquote>
<div>Eigen is a C++ template library for linear algebra: matrices,
vectors, numerical solvers, and related algorithms.</div></blockquote>
<p><a class="reference external" href="http://www.open-mpi.org">OpenMPI</a></p>
<blockquote>
<div>MPI is a standardized API typically used for parallel and/or
distributed computing. Open MPI is an open source, freely available
implementation.</div></blockquote>
<p><a class="reference external" href="http://www.gnu.org/software/gsl">GSL - GNU Scientific Library</a></p>
<blockquote>
<div>The GNU Scientific Library (GSL) is a numerical library for C and
C++ programmers.</div></blockquote>
<p><a class="reference external" href="http://gcc.gnu.org">GCC, the GNU Compiler</a></p>
<blockquote>
<div>The GNU Compiler Collection is a compiler system produced by the GNU
Project supporting various programming languages.</div></blockquote>
</div>
<div class="section" id="python-packages">
<h2>Python Packages<a class="headerlink" href="#python-packages" title="Permalink to this headline">¶</a></h2>
<p>To plot visualizations of the results, you will need Python 2.7 and the
packages listed below.</p>
<p><strong>Instructions:</strong> Install with <a class="reference external" href="http://www.pip-installer.org">pip</a>:</p>
<div class="highlight-python"><div class="highlight"><pre>pip install docopt numpy pandas matplotlib
</pre></div>
</div>
<p><strong>Note:</strong> The packages available on the Ubuntu repositories may be
outdated and might fail to work. So, avoid using <tt class="docutils literal"><span class="pre">apt-get</span></tt> for these
dependencies.</p>
<p><a class="reference external" href="http://docopt.org/">docopt</a></p>
<blockquote>
<div>Command-line interface description language.</div></blockquote>
<p><a class="reference external" href="http://www.numpy.org">numpy</a></p>
<blockquote>
<div>NumPy is the fundamental package for scientific computing with
Python.</div></blockquote>
<p><a class="reference external" href="http://pandas.pydata.org">pandas</a></p>
<blockquote>
<div>pandas is an open source, BSD-licensed library providing
high-performance, easy-to-use data structures and data analysis
tools for the Python programming language.</div></blockquote>
<p><a class="reference external" href="http://matplotlib.org">matplotlib</a></p>
<blockquote>
<div>matplotlib is a python 2D plotting library which produces
publication quality figures in a variety of hardcopy formats and
interactive environments across platforms.</div></blockquote>
<p><strong>Note:</strong> On a server with no display, please edit your
<a class="reference external" href="http://matplotlib.org/users/customizing.html">matplotlibrc</a> file to
use the <tt class="docutils literal"><span class="pre">Agg</span></tt> backend:</p>
<div class="highlight-python"><div class="highlight"><pre>perl -i -pe &#39;s/^(\s*(backend).*)$/#$1\n$2:Agg/&#39; ~/.matplotlib/matplotlibrc
</pre></div>
</div>
<p>Otherwise, you may see an error message like this:</p>
<div class="highlight-python"><div class="highlight"><pre>_tkinter.TclError: no display name and no $DISPLAY environment variable
</pre></div>
</div>
</div>
<div class="section" id="r-packages">
<h2>R Packages<a class="headerlink" href="#r-packages" title="Permalink to this headline">¶</a></h2>
<p>Some visualizations use R and ggplot2 instead of Python and matplotlib.</p>
<p><strong>Instructions:</strong> Start a session in R and run:</p>
<div class="code r highlight-python"><div class="highlight"><pre><span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s">&quot;data.table&quot;</span><span class="p">,</span> <span class="s">&quot;reshape2&quot;</span><span class="p">,</span> <span class="s">&quot;gap&quot;</span><span class="p">,</span> <span class="s">&quot;ggplot2&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p><a class="reference external" href="http://cran.r-project.org/web/packages/data.table">data.table</a></p>
<blockquote>
<div>Extension of data.frame for fast indexing, fast ordered joins, fast
assignment, fast grouping and list columns.</div></blockquote>
<p><a class="reference external" href="http://cran.r-project.org/web/packages/reshape2">reshape2</a></p>
<blockquote>
<div>Flexibly reshape data: a reboot of the reshape package.</div></blockquote>
<p><a class="reference external" href="http://cran.r-project.org/web/packages/gap">gap</a></p>
<blockquote>
<div>Genetic analysis package.</div></blockquote>
<p><a class="reference external" href="http://cran.r-project.org/web/packages/ggplot2">ggplot2</a></p>
<blockquote>
<div>An implementation of the Grammar of Graphics.</div></blockquote>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="data.html" class="btn btn-neutral float-right" title="Data"/>Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="algorithm.html" class="btn btn-neutral" title="Algorithm Details"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014, Kamil Slowikowski.
    </p>
  </div>

  <a href="https://github.com/snide/sphinx_rtd_theme">Sphinx theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.3',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>