

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Usage &mdash; SNPsea 1.0.3 documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="top" title="SNPsea 1.0.3 documentation" href="index.html"/>
        <link rel="next" title="Output Visualizations" href="output.html"/>
        <link rel="prev" title="Data" href="data.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        <a href="index.html" class="fa fa-home"> SNPsea</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
        
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="visual.html">Visual Summary</a><ul>
<li class="toctree-l2"><a class="reference internal" href="visual.html#cartoon">Cartoon</a></li>
<li class="toctree-l2"><a class="reference internal" href="visual.html#flow-chart">Flow Chart</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="algorithm.html">Algorithm Details</a><ul>
<li class="toctree-l2"><a class="reference internal" href="algorithm.html#step-1-assigning-genes-to-each-snp">Step 1: Assigning genes to each SNP</a></li>
<li class="toctree-l2"><a class="reference internal" href="algorithm.html#step-2-calculating-specificity-scores">Step 2: Calculating specificity scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="algorithm.html#step-3-testing-significance">Step 3: Testing significance</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="installation.html#c-libraries">C++ Libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#python-packages">Python Packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#r-packages">R Packages</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="data.html">Data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="data.html#snp-sets">SNP sets</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#gene-matrices">Gene matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#ld-pruned-snps-and-genomic-intervals">LD-pruned SNPs and Genomic Intervals</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#options">Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#input-file-formats">Input File Formats</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output-files">Output Files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="output.html">Output Visualizations</a><ul>
<li class="toctree-l2"><a class="reference internal" href="output.html#view-enrichment-of-tissue-specific-gene-expression">View enrichment of tissue-specific gene expression</a></li>
<li class="toctree-l2"><a class="reference internal" href="output.html#view-the-most-specifically-expressed-gene-for-each-snp-tissue-pair">View the most specifically expressed gene for each SNP-tissue pair</a></li>
<li class="toctree-l2"><a class="reference internal" href="output.html#view-the-type-1-error-rate-estimates-for-each-tissue">View the type 1 error rate estimates for each tissue</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="supplementary.html">Supplementary Figures</a><ul>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-1-determining-snp-linkage-intervals">Supplementary Figure 1: Determining SNP linkage intervals</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-2-counting-genes-in-gwas-snp-linkage-intervals">Supplementary Figure 2: Counting genes in GWAS SNP linkage intervals</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-3-choosing-the-threshold-for-linkage-intervals">Supplementary Figure 3: Choosing the <span class="math">\(r^{2}\)</span> threshold for linkage intervals</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-4-each-trait-associated-locus-harbors-a-single-associated-gene">Supplementary Figure 4: Each trait-associated locus harbors a single associated gene</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-5-type-1-error-estimates">Supplementary Figure 5: Type 1 error estimates</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-6-red-blood-cell-count-go-enrichment">Supplementary Figure 6: Red blood cell count GO enrichment</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-7-multiple-sclerosis">Supplementary Figure 7: Multiple sclerosis</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-8-celiac-disease">Supplementary Figure 8: Celiac disease</a></li>
<li class="toctree-l2"><a class="reference internal" href="supplementary.html#supplementary-figure-9-hdl-cholesterol">Supplementary Figure 9: HDL cholesterol</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">SNPsea</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Usage</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/usage.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            
  <div class="section" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<p>Here is a <a class="reference external" href="http://www.gnu.org/software/bash/manual/bashref.html">Bash</a>
script with a usage example:</p>
<div class="code bash highlight-python"><div class="highlight"><pre>options=(
    --snps              Red_blood_cell_count-Harst2012-45_SNPs.gwas
    --gene-matrix       GeneAtlas2004.gct.gz
    --gene-intervals    NCBIgenes2013.bed.gz
    --snp-intervals     TGP2011.bed.gz
    --null-snps         Lango2010.txt.gz
    --out               out
    --slop              10e3
    --threads           4
    --null-snpsets      0
    --min-observations  100
    --max-iterations    1e7
)
snpsea ${options[*]}
</pre></div>
</div>
<p>SNPsea will test SNPs associated with Red blood cell count for
tissue-specific expression of linked genes across 79 human tissues in
the Gene Atlas expression matrix. Each tissue will be tested up to 10
million times with matched random SNP sets, or testing will stop for a
tissue if 100 matched SNP sets achieve a higher specificity score than
the user&#8217;s SNPs.</p>
<div class="section" id="options">
<h2>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h2>
<p>All input files may optionally be compressed with
<tt class="docutils literal"><span class="pre">`gzip</span></tt> &lt;<a class="reference external" href="http://www.gzip.org/">http://www.gzip.org/</a>&gt;`__.</p>
<div class="section" id="required">
<h3>Required<a class="headerlink" href="#required" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>--snps ARG               Text file with SNP identifiers in the first
                         column. Instead of a file name, you may use
                         &#39;randomN&#39; with an integer N for a random SNP list
                         of length N.

--gene-matrix ARG        Gene matrix file in GCT format. The Name column
                         must contain the same gene identifiers as in
                         --gene-intervals.

--gene-intervals ARG     BED file with gene intervals. The fourth column
                         must contain the same gene identifiers as in
                         --gene-matrix.

--snp-intervals ARG      BED file with all known SNP intervals. The fourth
                         column must contain the same SNP identifiers as
                         in --snps and --null-snps.

--null-snps ARG          Text file with names of SNPs to sample when
                         generating null matched or random SNP sets.
                         These SNPs must be a subset of --snp-intervals.

--out ARG                Create output files in this directory. It will be
                         created if it does not already exist.
</pre></div>
</div>
</div>
<div class="section" id="optional">
<h3>Optional<a class="headerlink" href="#optional" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>--condition ARG          Text file with a list of columns in --gene-matrix
                         to condition on before calculating p-values. Each
                         column in --gene-matrix is projected onto each
                         column listed in this file and its projection is
                         subtracted.

--slop ARG               If a SNP interval overlaps no gene intervals,
                         extend the SNP interval this many nucleotides
                         further and try again.
                         [default: 10000]

--threads ARG            Number of threads to use.
                         [default: 1]

--null-snpsets ARG       Test this many null matched SNP sets, so you can
                         compare your results to a distribution of null
                         results.
                         [default: 0]

--min-observations ARG   Stop testing a column in --gene-matrix after
                         observing this many null SNP sets with
                         specificity scores greater or equal to those
                         obtained with the SNP set in --snps. Increase
                         this value to obtain more accurate p-values.
                         [default: 25]

--max-iterations ARG     Maximum number of null SNP sets tested for each
                         column in --gene-matrix. Increase this value to
                         resolve smaller p-values.
                         [default: 10000]
</pre></div>
</div>
</div>
</div>
<div class="section" id="input-file-formats">
<h2>Input File Formats<a class="headerlink" href="#input-file-formats" title="Permalink to this headline">¶</a></h2>
<div class="section" id="snps-arg">
<h3><tt class="docutils literal"><span class="pre">--snps</span> <span class="pre">ARG</span></tt><a class="headerlink" href="#snps-arg" title="Permalink to this headline">¶</a></h3>
<p>You must provide one or more comma-separated text files. SNP identifiers
must be listed one per line. Lines starting with <tt class="docutils literal"><span class="pre">#</span></tt> are skipped. If
the file has no header, the first column is assumed to contain SNP
identifiers. Otherwise, SNPsea looks for a column named (case-sensitive)
<tt class="docutils literal"><span class="pre">SNP</span></tt> or <tt class="docutils literal"><span class="pre">snp</span></tt> or <tt class="docutils literal"><span class="pre">name</span></tt> or <tt class="docutils literal"><span class="pre">marker</span></tt>.</p>
<div class="highlight-python"><div class="highlight"><pre>head Red_blood_cell_count-Harst2012-45_SNPs.gwas

# Harst et al. 2012
# doi:10.1038/nature11677
# PMID: 23222517
# 45 SNPs associated with red blood cell count (RBC) taken from Table 1.
# Positions are on hg19. SNPs are included if $P \le 5e-8$.
CHR POS SNP P
chr1    40069939    rs3916164   3e-10
chr1    158575729   rs857684    4e-16
chr1    199007208   rs7529925   8e-09
chr1    248039451   rs3811444   5e-10
</pre></div>
</div>
<p>Instead of providing a file with SNPs, you may use &#8220;randomN&#8221; like this:</p>
<div class="highlight-python"><div class="highlight"><pre>--snps random20
</pre></div>
</div>
<p>to sample 20 random SNPs from the <strong>``&#8211;snp-intervals``</strong> file.</p>
</div>
<div class="section" id="gene-matrix-arg">
<h3><tt class="docutils literal"><span class="pre">--gene-matrix</span> <span class="pre">ARG</span></tt><a class="headerlink" href="#gene-matrix-arg" title="Permalink to this headline">¶</a></h3>
<p>You must provide a single gene matrix that must be in
<a class="reference external" href="http://www.broadinstitute.org/cancer/software/genepattern/gp_guides/file-formats/sections/gct">GCT</a>
format.</p>
<div class="highlight-python"><div class="highlight"><pre>zcat GeneAtlas2004.gct.gz | cut -f1-4 | head

#1.2
17581  79
Name   Description  Colorectal_Adenocarcinoma  Whole_Blood
1      A1BG         115.5                      209.5
2      A2M          85                         328.5
9      NAT1         499                        1578
10     NAT2         115                        114
12     SERPINA3     419.5                      387.5
13     AADAC        125                        252.5
14     AAMP         2023                       942.5
</pre></div>
</div>
</div>
<div class="section" id="condition-arg-optional">
<h3><tt class="docutils literal"><span class="pre">--condition</span> <span class="pre">ARG</span></tt> (Optional)<a class="headerlink" href="#condition-arg-optional" title="Permalink to this headline">¶</a></h3>
<p>You may provide column names present in the <strong>``&#8211;gene-matrix``</strong> file,
one per line. The matrix will be conditioned on these columns before the
analysis is performed to help you identify secondary signals independent
of these columns. Binary (0, 1) matrices will not be conditioned.</p>
<div class="highlight-python"><div class="highlight"><pre>head conditions.txt

Whole_Blood
</pre></div>
</div>
</div>
<div class="section" id="gene-intervals-arg">
<h3><tt class="docutils literal"><span class="pre">--gene-intervals</span> <span class="pre">ARG</span></tt><a class="headerlink" href="#gene-intervals-arg" title="Permalink to this headline">¶</a></h3>
<p>You must provide gene intervals in BED format with a fourth column that
contains the same gene identifiers as those present in the Name column
of the <strong>``&#8211;gene-matrix``</strong>
<a class="reference external" href="http://www.broadinstitute.org/cancer/software/genepattern/gp_guides/file-formats/sections/gct">GCT</a>
file. Only the first four columns are used.</p>
<div class="highlight-python"><div class="highlight"><pre>zcat NCBIgenes2013.bed.gz | head

chr1  10003485   10045555   64802      NMNAT1
chr1  100111430  100160096  54873      PALMD
chr1  100163795  100164756  100129320  HMGB3P10
chr1  100174205  100232185  391059     FRRS1
chr1  10027438   10027515   100847055  MIR5697
chr1  100308165  100308317  100270894  RPL39P9
chr1  100315632  100389578  178        AGL
chr1  100433941  100435837  730081     LOC730081
chr1  100435344  100492534  23443      SLC35A3
chr1  100503669  100548932  64645      HIAT1
</pre></div>
</div>
</div>
<div class="section" id="snp-intervals-arg">
<h3><tt class="docutils literal"><span class="pre">--snp-intervals</span> <span class="pre">ARG</span></tt><a class="headerlink" href="#snp-intervals-arg" title="Permalink to this headline">¶</a></h3>
<p>SNP linkage intervals must be specified in BED format and include a
fourth column with the SNP identifiers. The linkage intervals assigned
to the trait-associated SNPs you provide with <strong>``&#8211;snps``</strong> are taken
from this file.</p>
<div class="highlight-python"><div class="highlight"><pre>zcat TGP2011.bed.gz | head

chr1    0   254996  rs113759966
chr1    0   254996  rs114420996
chr1    0   254996  rs114608975
chr1    0   254996  rs115209712
chr1    0   254996  rs116400033
chr1    0   254996  rs116504101
chr1    0   254996  rs12184306
chr1    0   254996  rs12184307
chr1    0   254996  rs138808727
chr1    0   254996  rs139113303
</pre></div>
</div>
</div>
<div class="section" id="null-snps-arg">
<h3><tt class="docutils literal"><span class="pre">--null-snps</span> <span class="pre">ARG</span></tt><a class="headerlink" href="#null-snps-arg" title="Permalink to this headline">¶</a></h3>
<p>The null SNPs file must have one SNP identifier per line. Only the first
column is used. The identifiers must be a subset of the identifiers in
<strong>``&#8211;snp-intervals``</strong>.</p>
<div class="highlight-python"><div class="highlight"><pre>zcat Lango2010.txt.gz | head

rs58108140  chr1    10583
rs180734498 chr1    13302
rs140337953 chr1    30923
rs141149254 chr1    54490
rs2462492   chr1    54676
rs10399749  chr1    55299
rs189727433 chr1    57952
rs149755937 chr1    59040
rs77573425  chr1    61989
rs116440577 chr1    63671
</pre></div>
</div>
</div>
</div>
<div class="section" id="output-files">
<h2>Output Files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>The usage example shown above produces the following output files:</p>
<div class="highlight-python"><div class="highlight"><pre>out/
    args.txt
    condition_pvalues.txt
    null_pvalues.txt
    snp_condition_scores.txt
    snp_genes.txt
</pre></div>
</div>
<div class="section" id="args-txt">
<h3><tt class="docutils literal"><span class="pre">args.txt</span></tt><a class="headerlink" href="#args-txt" title="Permalink to this headline">¶</a></h3>
<p>The command line arguments needed to reproduce the analysis.</p>
<div class="highlight-python"><div class="highlight"><pre>cat args.txt

# SNPsea v1.0.2
--snps             Red_blood_cell_count-Harst2012-45_SNPs.gwas
--gene-matrix      GeneAtlas2004.gct.gz
--gene-intervals   NCBIgenes2013.bed.gz
--snp-intervals    TGP2011.bed.gz
--null-snps        Lango2010.txt.gz
--out              out
--score            single
--slop             100000
--threads          8
--null-snpsets     0
--min-observations 100
--max-iterations   10000000
</pre></div>
</div>
<p>Repeat the analysis:</p>
<div class="highlight-python"><div class="highlight"><pre>snpsea --args args.txt
</pre></div>
</div>
</div>
<div class="section" id="condition-pvalues-txt">
<h3><tt class="docutils literal"><span class="pre">condition_pvalues.txt</span></tt><a class="headerlink" href="#condition-pvalues-txt" title="Permalink to this headline">¶</a></h3>
<p>The p-values representing enrichment of condition-specificity for the
given SNPs.</p>
<div class="highlight-python"><div class="highlight"><pre>head condition_pvalues.txt | column -t

condition                  pvalue     nulls_observed  nulls_tested
Colorectal_Adenocarcinoma  0.933555   280             300
Whole_Blood                0.521595   156             300
BM-CD33+Myeloid            0.159772   111             700
PB-CD14+Monocytes          0.103264   154             1500
PB-BDCA4+Dentritic_cells   0.0606256  187             3100
PB-CD56+NK_cells           0.194009   135             700
PB-CD4+T_cells             0.428571   128             300
PB-CD8+T_cells             0.531561   159             300
PB-CD19+B_cells            0.226819   158             700
</pre></div>
</div>
</div>
<div class="section" id="null-pvalues-txt">
<h3><tt class="docutils literal"><span class="pre">null_pvalues.txt</span></tt><a class="headerlink" href="#null-pvalues-txt" title="Permalink to this headline">¶</a></h3>
<p>If the argument for <strong>``&#8211;snps``</strong> is the name of a file, the p-values
for null matched SNP sets. You can compare these null results to the
results for your trait-associated SNPs.</p>
<p>If the argument for <strong>``&#8211;snps``</strong> is &#8220;randomN&#8221; where N is some integer,
like &#8220;random20&#8221; the p-values for random unmatched SNP sets, each with N
SNPs.</p>
<p>The fifth column is the replicate index. The number of replicates
performed is specified with <strong>``&#8211;null-snpsets INT``</strong>.</p>
<div class="highlight-python"><div class="highlight"><pre>head null_pvalues.txt | column -t

ColorectalAdenocarcinoma  0.056     84   1500  0
WholeBlood                0.236667  71   300   0
BM-CD33+Myeloid           0.55      55   100   0
PB-CD14+Monocytes         0.59      59   100   0
PB-BDCA4+Dentritic_Cells  0.59      59   100   0
PB-CD56+NKCells           0.71      71   100   0
PB-CD4+Tcells             0.383333  115  300   0
PB-CD8+Tcells             0.128571  90   700   0
PB-CD19+Bcells            0.168571  118  700   0
BM-CD105+Endothelial      0.386667  116  300   0
</pre></div>
</div>
</div>
<div class="section" id="snp-genes-txt">
<h3><tt class="docutils literal"><span class="pre">snp_genes.txt</span></tt><a class="headerlink" href="#snp-genes-txt" title="Permalink to this headline">¶</a></h3>
<p>Each SNP&#8217;s linkage interval and overlapping genes. If a SNP is not found
in the reference file specified with <strong>``&#8211;snp-intervals``</strong>, then the
name of the SNP will be listed and the other columns will contain
<tt class="docutils literal"><span class="pre">NA</span></tt>.</p>
<div class="highlight-python"><div class="highlight"><pre>head snp_genes.txt | column -t

chrom  start      end        snp         n_genes  genes
chr4   55364224   55408999   rs218238    0        NA
chr6   139827777  139844854  rs590856    0        NA
NA     NA         NA         rs99999999  NA       NA
chr6   109505894  109651220  rs1008084   2        8763,27244
chr10  71089843   71131638   rs10159477  1        3098
chr2   111807303  111856057  rs10207392  1        55289
chr16  88831494   88903796   rs10445033  4        353,2588,9780,81620
chr7   151396253  151417368  rs10480300  1        51422
chr12  4320955    4336783    rs10849023  2        894,57103
chr15  76129642   76397903   rs11072566  4        26263,92912,123591,145957
</pre></div>
</div>
</div>
<div class="section" id="snp-condition-scores-txt">
<h3><tt class="docutils literal"><span class="pre">snp_condition_scores.txt</span></tt><a class="headerlink" href="#snp-condition-scores-txt" title="Permalink to this headline">¶</a></h3>
<p>Each SNP, condition, gene with greatest specificity to that condition,
and score for the SNP-condition pair, adjusted for the number of genes
overlapping the given SNP&#8217;s linkage interval.</p>
<div class="highlight-python"><div class="highlight"><pre>head snp_condition_scores.txt | column -t

snp        condition                  gene   score
rs9349204  Colorectal_Adenocarcinoma  10817  0.693027
rs9349204  Whole_Blood                896    0.285864
rs9349204  BM-CD33+Myeloid            896    0.236487
rs9349204  PB-CD14+Monocytes          29964  0.340561
rs9349204  PB-BDCA4+Dentritic_cells   29964  0.411727
rs9349204  PB-CD56+NK_cells           896    0.0356897
rs9349204  PB-CD4+T_cells             896    0.38182
rs9349204  PB-CD8+T_cells             896    0.332008
rs9349204  PB-CD19+B_cells            29964  0.255196
</pre></div>
</div>
</div>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="output.html" class="btn btn-neutral float-right" title="Output Visualizations"/>Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="data.html" class="btn btn-neutral" title="Data"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014, Kamil Slowikowski.
    </p>
  </div>

  <a href="https://github.com/snide/sphinx_rtd_theme">Sphinx theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.3',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>